// @traqula/rules-sparql-1-1/dist/esm/lib/MinimalSparqlParser.js
import { traqulaNewlineAlternative, TransformerObject, traqulaIndentation } from '@traqula/core';
import { AstFactory } from './astFactory.js';
export function completeParseContext(context) {
    return {
        astFactory: context.astFactory ?? new AstFactory({ tracksSourceLocation: false }),
        baseIRI: context.baseIRI,
        prefixes: { ...context.prefixes },
        parseMode: context.parseMode ? new Set(context.parseMode) : new Set(['canParseVars', 'canCreateBlankNodes']),
        skipValidation: context.skipValidation ?? false,
    };
}
export function completeGeneratorContext(context) {
    return {
        astFactory: context.astFactory ?? new AstFactory(),
        origSource: context.origSource ?? '',
        offset: context.offset,
        [traqulaIndentation]: context[traqulaIndentation] ?? 0,
        [traqulaNewlineAlternative]: context[traqulaNewlineAlternative] ?? ' ',
        indentInc: context.indentInc ?? 2,
    };
}
export function copyParseContext(context) {
    return {
        ...context,
        prefixes: { ...context.prefixes },
        parseMode: new Set(context.parseMode),
    };
}
export class MinimalSparqlParser {
    parser;
    defaultContext;
    coreTransformer = new TransformerObject();
    constructor(parser, defaultContext = {}) {
        this.parser = parser;
        this.defaultContext = completeParseContext(defaultContext);
    }
    /**
     * Parse a query string starting from the
     * [QueryUnit](https://www.w3.org/TR/sparql11-query/#rQueryUnit)
     * or [QueryUpdate](https://www.w3.org/TR/sparql11-query/#rUpdateUnit) rules.
     * @param query
     * @param context
     */
    parse(query, context = {}) {
        const ast = this.parser.queryOrUpdate(query, copyParseContext({ ...this.defaultContext, ...context }));
        ast.loc = this.defaultContext.astFactory.sourceLocationInlinedSource(query, ast.loc, 0, Number.MAX_SAFE_INTEGER);
        return ast;
    }
    /**
     * Parse a query string starting from the [Path](https://www.w3.org/TR/sparql11-query/#rPath) grammar rule.
     * @param query
     * @param context
     */
    parsePath(query, context = {}) {
        const ast = this.parser.path(query, copyParseContext({ ...this.defaultContext, ...context }));
        ast.loc = this.defaultContext.astFactory.sourceLocationInlinedSource(query, ast.loc, 0, Number.MAX_SAFE_INTEGER);
        if (this.defaultContext.astFactory.isPathPure(ast)) {
            return {
                ...ast,
                prefixes: {},
            };
        }
        return ast;
    }
}
