// @traqula/rules-sparql-1-1/dist/cjs/lib/lexer/BuiltInCalls.js
"use strict";
/* eslint-disable require-unicode-regexp */
Object.defineProperty(exports, "__esModule", { value: true });
exports.isnumeric = exports.isliteral = exports.isblank = exports.isuri = exports.isiri = exports.sameterm = exports.strdt = exports.strlang = exports.if_ = exports.coalesce = exports.sha512 = exports.sha384 = exports.sha256 = exports.sha1 = exports.md5 = exports.struuid = exports.uuid = exports.now = exports.tz = exports.timezone = exports.seconds = exports.minutes = exports.hours = exports.day = exports.month = exports.year = exports.strafter = exports.strbefore = exports.strends = exports.strstarts = exports.contains = exports.encode_for_uri = exports.lcase = exports.ucase = exports.strlen = exports.concat = exports.round = exports.floor = exports.ceil = exports.abs = exports.rand = exports.bnode = exports.uri = exports.iri = exports.bound = exports.datatype = exports.langmatches = exports.lang = exports.str = exports.BuiltInCalls = void 0;
exports.allBuiltInCalls = exports.groupConcat = exports.sample = exports.avg = exports.max = exports.min = exports.sum = exports.count = exports.notexists = exports.exists = exports.replace = exports.substr = exports.regex = void 0;
const core_1 = require("@traqula/core");
var BuiltInCalls;
(function (BuiltInCalls) {
    BuiltInCalls["Str"] = "builtInStr";
    BuiltInCalls["Lang"] = "builtInLang";
    BuiltInCalls["Langmatches"] = "builtInLangmatches";
    BuiltInCalls["Datatype"] = "builtInDatatype";
    BuiltInCalls["Bound"] = "builtInBound";
    BuiltInCalls["Iri"] = "builtInIri";
    BuiltInCalls["Uri"] = "builtInUri";
    BuiltInCalls["Bnode"] = "builtInBnode";
    BuiltInCalls["Rand"] = "builtInRand";
    BuiltInCalls["Abs"] = "builtInAbs";
    BuiltInCalls["Ceil"] = "builtInCeil";
    BuiltInCalls["Floor"] = "builtInFloor";
    BuiltInCalls["Round"] = "builtInRound";
    BuiltInCalls["Concat"] = "builtInConcat";
    BuiltInCalls["Strlen"] = "builtInStrlen";
    BuiltInCalls["Ucase"] = "builtInUcase";
    BuiltInCalls["Lcase"] = "builtInLcase";
    BuiltInCalls["Encode_for_uri"] = "builtInEncode_for_uri";
    BuiltInCalls["Contains"] = "builtInContains";
    BuiltInCalls["Strstarts"] = "builtInStrstarts";
    BuiltInCalls["Strends"] = "builtInStrends";
    BuiltInCalls["Strbefore"] = "builtInStrbefore";
    BuiltInCalls["Strafter"] = "builtInStrafter";
    BuiltInCalls["Year"] = "builtInYear";
    BuiltInCalls["Month"] = "builtInMonth";
    BuiltInCalls["Day"] = "builtInDay";
    BuiltInCalls["Hours"] = "builtInHours";
    BuiltInCalls["Minutes"] = "builtInMinutes";
    BuiltInCalls["Seconds"] = "builtInSeconds";
    BuiltInCalls["Timezone"] = "builtInTimezone";
    BuiltInCalls["Tz"] = "builtInTz";
    BuiltInCalls["Now"] = "builtInNow";
    BuiltInCalls["Uuid"] = "builtInUuid";
    BuiltInCalls["Struuid"] = "builtInStruuid";
    BuiltInCalls["Md5"] = "builtInMd5";
    BuiltInCalls["Sha1"] = "builtInSha1";
    BuiltInCalls["Sha256"] = "builtInSha256";
    BuiltInCalls["Sha384"] = "builtInSha384";
    BuiltInCalls["Sha512"] = "builtInSha512";
    BuiltInCalls["Coalesce"] = "builtInCoalesce";
    BuiltInCalls["If"] = "builtInIf";
    BuiltInCalls["Strlang"] = "builtInStrlang";
    BuiltInCalls["Strdt"] = "builtInStrdt";
    BuiltInCalls["Sameterm"] = "builtInSameterm";
    BuiltInCalls["Isiri"] = "builtInIsiri";
    BuiltInCalls["Isuri"] = "builtInIsuri";
    BuiltInCalls["Isblank"] = "builtInIsblank";
    BuiltInCalls["Isliteral"] = "builtInIsliteral";
    BuiltInCalls["Isnumeric"] = "builtInIsnumeric";
    BuiltInCalls["Regex"] = "builtInRegex";
    BuiltInCalls["Substr"] = "builtInSubstr";
    BuiltInCalls["Replace"] = "builtInReplace";
    BuiltInCalls["Exists"] = "builtInExists";
    BuiltInCalls["Notexists"] = "builtInNotexists";
    BuiltInCalls["Count"] = "builtInCount";
    BuiltInCalls["Sum"] = "builtInSum";
    BuiltInCalls["Min"] = "builtInMin";
    BuiltInCalls["Max"] = "builtInMax";
    BuiltInCalls["Avg"] = "builtInAvg";
    BuiltInCalls["Sample"] = "builtInSample";
    BuiltInCalls["Group_concat"] = "builtInGroup_concat";
})(BuiltInCalls || (exports.BuiltInCalls = BuiltInCalls = {}));
function capitalize(string) {
    return (string.charAt(0).toUpperCase() + string.slice(1));
}
exports.str = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Str), pattern: /str/i, label: 'STR' });
exports.lang = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Lang), pattern: /lang/i, label: 'LANG' });
exports.langmatches = (0, core_1.createToken)({
    name: capitalize(BuiltInCalls.Langmatches),
    pattern: /langmatches/i,
    label: 'LANGMATCHES',
});
exports.datatype = (0, core_1.createToken)({
    name: capitalize(BuiltInCalls.Datatype),
    pattern: /datatype/i,
    label: 'DATATYPE',
});
exports.bound = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Bound), pattern: /bound/i, label: 'BOUND' });
exports.iri = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Iri), pattern: /iri/i, label: 'IRI' });
exports.uri = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Uri), pattern: /uri/i, label: 'URI' });
exports.bnode = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Bnode), pattern: /bnode/i, label: 'BNODE' });
exports.rand = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Rand), pattern: /rand/i, label: 'RAND' });
exports.abs = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Abs), pattern: /abs/i, label: 'ABS' });
exports.ceil = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Ceil), pattern: /ceil/i, label: 'CEIL' });
exports.floor = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Floor), pattern: /floor/i, label: 'FLOOR' });
exports.round = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Round), pattern: /round/i, label: 'ROUND' });
exports.concat = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Concat), pattern: /concat/i, label: 'CONCAT' });
exports.strlen = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Strlen), pattern: /strlen/i, label: 'STRLEN' });
exports.ucase = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Ucase), pattern: /ucase/i, label: 'UCASE' });
exports.lcase = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Lcase), pattern: /lcase/i, label: 'LCASE' });
exports.encode_for_uri = (0, core_1.createToken)({
    name: capitalize(BuiltInCalls.Encode_for_uri),
    pattern: /encode_for_uri/i,
    label: 'ENCODE_FOR_URI',
});
exports.contains = (0, core_1.createToken)({
    name: capitalize(BuiltInCalls.Contains),
    pattern: /contains/i,
    label: 'CONTAINS',
});
exports.strstarts = (0, core_1.createToken)({
    name: capitalize(BuiltInCalls.Strstarts),
    pattern: /strstarts/i,
    label: 'STRSTARTS',
});
exports.strends = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Strends), pattern: /strends/i, label: 'STRENDS' });
exports.strbefore = (0, core_1.createToken)({
    name: capitalize(BuiltInCalls.Strbefore),
    pattern: /strbefore/i,
    label: 'STRBEFORE',
});
exports.strafter = (0, core_1.createToken)({
    name: capitalize(BuiltInCalls.Strafter),
    pattern: /strafter/i,
    label: 'STRAFTER',
});
exports.year = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Year), pattern: /year/i, label: 'YEAR' });
exports.month = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Month), pattern: /month/i, label: 'MONTH' });
exports.day = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Day), pattern: /day/i, label: 'DAY' });
exports.hours = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Hours), pattern: /hours/i, label: 'HOURS' });
exports.minutes = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Minutes), pattern: /minutes/i, label: 'MINUTES' });
exports.seconds = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Seconds), pattern: /seconds/i, label: 'SECONDS' });
exports.timezone = (0, core_1.createToken)({
    name: capitalize(BuiltInCalls.Timezone),
    pattern: /timezone/i,
    label: 'TIMEZONE',
});
exports.tz = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Tz), pattern: /tz/i, label: 'TZ' });
exports.now = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Now), pattern: /now/i, label: 'NOW' });
exports.uuid = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Uuid), pattern: /uuid/i, label: 'UUID' });
exports.struuid = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Struuid), pattern: /struuid/i, label: 'STRUUID' });
exports.md5 = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Md5), pattern: /md5/i, label: 'MD5' });
exports.sha1 = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Sha1), pattern: /sha1/i, label: 'SHA1' });
exports.sha256 = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Sha256), pattern: /sha256/i, label: 'SHA256' });
exports.sha384 = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Sha384), pattern: /sha384/i, label: 'SHA384' });
exports.sha512 = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Sha512), pattern: /sha512/i, label: 'SHA512' });
exports.coalesce = (0, core_1.createToken)({
    name: capitalize(BuiltInCalls.Coalesce),
    pattern: /coalesce/i,
    label: 'COALESCE',
});
exports.if_ = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.If), pattern: /if/i, label: 'IF' });
exports.strlang = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Strlang), pattern: /strlang/i, label: 'STRLANG' });
exports.strdt = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Strdt), pattern: /strdt/i, label: 'STRDT' });
exports.sameterm = (0, core_1.createToken)({
    name: capitalize(BuiltInCalls.Sameterm),
    pattern: /sameterm/i,
    label: 'SAMETERM',
});
exports.isiri = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Isiri), pattern: /isiri/i, label: 'ISIRI' });
exports.isuri = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Isuri), pattern: /isuri/i, label: 'ISURI' });
exports.isblank = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Isblank), pattern: /isblank/i, label: 'ISBLANK' });
exports.isliteral = (0, core_1.createToken)({
    name: capitalize(BuiltInCalls.Isliteral),
    pattern: /isliteral/i,
    label: 'ISLITERAL',
});
exports.isnumeric = (0, core_1.createToken)({
    name: capitalize(BuiltInCalls.Isnumeric),
    pattern: /isnumeric/i,
    label: 'ISNUMERIC',
});
exports.regex = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Regex), pattern: /regex/i, label: 'REGEX' });
exports.substr = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Substr), pattern: /substr/i, label: 'SUBSTR' });
exports.replace = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Replace), pattern: /replace/i, label: 'REPLACE' });
exports.exists = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Exists), pattern: /exists/i, label: 'EXISTS' });
exports.notexists = (0, core_1.createToken)({
    name: capitalize(BuiltInCalls.Notexists),
    pattern: /not exists/i,
    label: 'NOT EXISTS',
});
exports.count = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Count), pattern: /count/i, label: 'COUNT' });
exports.sum = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Sum), pattern: /sum/i, label: 'SUM' });
exports.min = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Min), pattern: /min/i, label: 'MIN' });
exports.max = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Max), pattern: /max/i, label: 'MAX' });
exports.avg = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Avg), pattern: /avg/i, label: 'AVG' });
exports.sample = (0, core_1.createToken)({ name: capitalize(BuiltInCalls.Sample), pattern: /sample/i, label: 'SAMPLE' });
exports.groupConcat = (0, core_1.createToken)({
    name: capitalize(BuiltInCalls.Group_concat),
    pattern: /group_concat/i,
    label: 'GROUP_CONCAT',
});
exports.allBuiltInCalls = core_1.LexerBuilder.create().add(exports.langmatches, exports.datatype, exports.lang, exports.bound, exports.iri, exports.uri, exports.bnode, exports.rand, exports.abs, exports.ceil, exports.floor, exports.round, exports.concat, exports.strlen, exports.ucase, exports.lcase, exports.encode_for_uri, exports.contains, exports.strstarts, exports.strends, exports.strbefore, exports.strafter, exports.year, exports.month, exports.day, exports.hours, exports.minutes, exports.seconds, exports.timezone, exports.tz, exports.now, exports.uuid, exports.struuid, exports.md5, exports.sha1, exports.sha256, exports.sha384, exports.sha512, exports.coalesce, exports.if_, exports.strlang, exports.strdt, exports.sameterm, exports.isiri, exports.isuri, exports.isblank, exports.isliteral, exports.isnumeric, exports.regex, exports.substr, exports.replace, exports.exists, exports.notexists, exports.count, exports.sum, exports.min, exports.max, exports.avg, exports.sample, exports.groupConcat, exports.str);
