// @traqula/rules-sparql-1-1/dist/cjs/lib/lexer/lexer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.separator = exports.notIn = exports.in_ = exports.false_ = exports.true_ = exports.a = exports.filter = exports.union = exports.minus = exports.undef = exports.bind = exports.service = exports.optional = exports.usingClause = exports.insertClause = exports.insertDataClause = exports.deleteClause = exports.deleteWhereClause = exports.deleteDataClause = exports.modifyWith = exports.copy = exports.move = exports.to = exports.add = exports.create = exports.drop = exports.clear = exports.loadInto = exports.silent = exports.load = exports.values = exports.offset = exports.limit = exports.orderDesc = exports.orderAsc = exports.order = exports.having = exports.by = exports.groupByGroup = exports.where = exports.from = exports.ask = exports.describe = exports.construct = exports.as = exports.reduced = exports.distinct = exports.select = exports.prefixDecl = exports.baseDecl = void 0;
exports.sparql11LexerBuilder = exports.allBaseTokens = void 0;
/* eslint-disable require-unicode-regexp */
const core_1 = require("@traqula/core");
const BuiltInCalls_js_1 = require("./BuiltInCalls.js");
const graph_js_1 = require("./graph.js");
const lexerPatterns_js_1 = require("./lexerPatterns.js");
const symbols_js_1 = require("./symbols.js");
const terminals_js_1 = require("./terminals.js");
exports.baseDecl = (0, core_1.createToken)({ name: 'BaseDecl', pattern: /base/i, label: 'BASE' });
exports.prefixDecl = (0, core_1.createToken)({ name: 'PrefixDecl', pattern: /prefix/i, label: 'PREFIX' });
exports.select = (0, core_1.createToken)({ name: 'Select', pattern: /select/i, label: 'SELECT' });
exports.distinct = (0, core_1.createToken)({ name: 'Distinct', pattern: /distinct/i, label: 'DISTINCT' });
exports.reduced = (0, core_1.createToken)({ name: 'Reduced', pattern: /reduced/i, label: 'REDUCED' });
exports.as = (0, core_1.createToken)({ name: 'As', pattern: /as/i, label: 'AS' });
exports.construct = (0, core_1.createToken)({ name: 'Construct', pattern: /construct/i, label: 'CONSTRUCT' });
exports.describe = (0, core_1.createToken)({ name: 'Describe', pattern: /describe/i, label: 'DESCRIBE' });
exports.ask = (0, core_1.createToken)({ name: 'Ask', pattern: /ask/i, label: 'ASK' });
exports.from = (0, core_1.createToken)({ name: 'From', pattern: /from/i, label: 'FROM' });
exports.where = (0, core_1.createToken)({ name: 'Where', pattern: /where/i, label: 'WHERE' });
exports.groupByGroup = (0, core_1.createToken)({ name: 'GroupByGroup', pattern: /group/i, label: '_GROUP_ BY' });
exports.by = (0, core_1.createToken)({ name: 'By', pattern: /by/i, label: 'BY' });
exports.having = (0, core_1.createToken)({ name: 'Having', pattern: /having/i, label: 'HAVING' });
exports.order = (0, core_1.createToken)({ name: 'Order', pattern: /order/i, label: '_ORDER_ BY' });
exports.orderAsc = (0, core_1.createToken)({ name: 'OrderAsc', pattern: /asc/i, label: 'ASC' });
exports.orderDesc = (0, core_1.createToken)({ name: 'OrderDesc', pattern: /desc/i, label: 'DESC' });
exports.limit = (0, core_1.createToken)({ name: 'Limit', pattern: /limit/i, label: 'LIMIT' });
exports.offset = (0, core_1.createToken)({ name: 'Offset', pattern: /offset/i, label: 'OFFSET' });
exports.values = (0, core_1.createToken)({ name: 'Values', pattern: /values/i, label: 'VALUES' });
exports.load = (0, core_1.createToken)({ name: 'Load', pattern: /load/i, label: 'LOAD' });
exports.silent = (0, core_1.createToken)({ name: 'Silent', pattern: /silent/i, label: 'SILENT' });
exports.loadInto = (0, core_1.createToken)({ name: 'LoadInto', pattern: /into/i, label: 'INTO' });
exports.clear = (0, core_1.createToken)({ name: 'Clear', pattern: /clear/i, label: 'CLEAR' });
exports.drop = (0, core_1.createToken)({ name: 'Drop', pattern: /drop/i, label: 'DROP' });
exports.create = (0, core_1.createToken)({ name: 'Create', pattern: /create/i, label: 'CREATE' });
exports.add = (0, core_1.createToken)({ name: 'Add', pattern: /add/i, label: 'ADD' });
exports.to = (0, core_1.createToken)({ name: 'To', pattern: /to/i, label: 'TO' });
exports.move = (0, core_1.createToken)({ name: 'Move', pattern: /move/i, label: 'MOVE' });
exports.copy = (0, core_1.createToken)({ name: 'Copy', pattern: /copy/i, label: 'COPY' });
exports.modifyWith = (0, core_1.createToken)({ name: 'ModifyWith', pattern: /with/i, label: 'WITH' });
exports.deleteDataClause = (0, core_1.createToken)({
    name: 'DeleteDataClause',
    pattern: new RegExp(`delete(${lexerPatterns_js_1.atLeastOneBlankPattern.source})data`, 'i'),
    label: 'DELETE DATA',
});
exports.deleteWhereClause = (0, core_1.createToken)({
    name: 'DeleteWhereClause',
    pattern: new RegExp(`delete(${lexerPatterns_js_1.atLeastOneBlankPattern.source})where`, 'i'),
    label: 'DELETE WHERE',
});
exports.deleteClause = (0, core_1.createToken)({ name: 'DeleteClause', pattern: /delete/i, label: 'DELETE' });
exports.insertDataClause = (0, core_1.createToken)({
    name: 'InsertDataClause',
    pattern: new RegExp(`insert(${lexerPatterns_js_1.atLeastOneBlankPattern.source})data`, 'i'),
    label: 'INSERT DATA',
});
exports.insertClause = (0, core_1.createToken)({ name: 'InsertClause', pattern: /insert/i, label: 'insert' });
exports.usingClause = (0, core_1.createToken)({ name: 'UsingClause', pattern: /using/i, label: 'USING' });
exports.optional = (0, core_1.createToken)({ name: 'Optional', pattern: /optional/i, label: 'OPTIONAL' });
exports.service = (0, core_1.createToken)({ name: 'Service', pattern: /service/i, label: 'SERVICE' });
exports.bind = (0, core_1.createToken)({ name: 'Bind', pattern: /bind/i, label: 'BIND' });
exports.undef = (0, core_1.createToken)({ name: 'Undef', pattern: /undef/i, label: 'UNDEF' });
exports.minus = (0, core_1.createToken)({ name: 'Minus', pattern: /minus/i, label: 'MINUS' });
exports.union = (0, core_1.createToken)({ name: 'Union', pattern: /union/i, label: 'UNION' });
exports.filter = (0, core_1.createToken)({ name: 'Filter', pattern: /filter/i, label: 'FILTER' });
exports.a = (0, core_1.createToken)({ name: 'a', pattern: 'a', label: 'type declaration \'a\'' });
exports.true_ = (0, core_1.createToken)({ name: 'True', pattern: /true/i, label: 'true' });
exports.false_ = (0, core_1.createToken)({ name: 'False', pattern: /false/i, label: 'false' });
exports.in_ = (0, core_1.createToken)({ name: 'In', pattern: /in/i, label: 'IN' });
// eslint-disable-next-line unicorn/better-regex,no-control-regex
exports.notIn = (0, core_1.createToken)({ name: 'NotIn', pattern: /not[\u0020\u0009\u000D\u000A]+in/i, label: 'NOT IN' });
exports.separator = (0, core_1.createToken)({ name: 'Separator', pattern: /separator/i, label: 'SEPARATOR' });
exports.allBaseTokens = core_1.LexerBuilder.create().add(exports.baseDecl, exports.prefixDecl, exports.select, exports.distinct, exports.reduced, exports.construct, exports.describe, exports.ask, exports.from, exports.where, exports.having, exports.groupByGroup, exports.by, exports.order, exports.orderAsc, exports.orderDesc, exports.limit, exports.offset, exports.values, exports.load, exports.silent, exports.loadInto, exports.clear, exports.drop, exports.create, exports.add, exports.to, exports.move, exports.copy, exports.modifyWith, exports.deleteWhereClause, exports.deleteDataClause, exports.deleteClause, exports.insertDataClause, exports.insertClause, exports.usingClause, exports.optional, exports.service, exports.bind, exports.undef, exports.minus, exports.union, exports.filter, exports.as, exports.a, exports.true_, exports.false_, exports.in_, exports.notIn, exports.separator);
/**
 * [!!!ORDER MATTERS!!!](https://chevrotain.io/docs/tutorial/step1_lexing.html#creating-the-lexer)
 */
exports.sparql11LexerBuilder = core_1.LexerBuilder
    .create(terminals_js_1.allTerminals)
    .merge(exports.allBaseTokens)
    .merge(BuiltInCalls_js_1.allBuiltInCalls)
    .merge(graph_js_1.allGraphTokens)
    .merge(symbols_js_1.allSymbols)
    .moveAfter(BuiltInCalls_js_1.avg, exports.a)
    .moveBefore(exports.a, graph_js_1.graphAll)
    .moveAfter(BuiltInCalls_js_1.groupConcat, exports.groupByGroup);
