// @traqula/rules-sparql-1-1/dist/cjs/lib/lexer/terminals.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.allTerminals = exports.anon = exports.nil = exports.comment = exports.ws = exports.stringLiteralLong2 = exports.stringLiteralLong1 = exports.stringLiteral2 = exports.stringLiteral1 = exports.doubleNegative = exports.decimalNegative = exports.integerNegative = exports.doublePositive = exports.decimalPositive = exports.integerPositive = exports.double = exports.decimal = exports.integer = exports.langTag = exports.var2 = exports.var1 = exports.blankNodeLabel = exports.pNameNs = exports.pNameLn = exports.iriRef = void 0;
const chevrotain_1 = require("@traqula/chevrotain");
const core_1 = require("@traqula/core");
const lexerPatterns_js_1 = require("./lexerPatterns.js");
/**
 * [[139]](https://www.w3.org/TR/sparql11-query/#rIRIREF)
 */
exports.iriRef = (0, core_1.createToken)({ name: 'IriRef', pattern: lexerPatterns_js_1.iriRefPattern });
/**
 * [[141]](https://www.w3.org/TR/sparql11-query/#rPNAME_LN)
 */
exports.pNameLn = (0, core_1.createToken)({ name: 'PNameLn', pattern: lexerPatterns_js_1.pNameLnPattern });
/**
 * [[140]](https://www.w3.org/TR/sparql11-query/#rPNAME_NS)
 */
exports.pNameNs = (0, core_1.createToken)({ name: 'PNameNs', pattern: lexerPatterns_js_1.pNameNsPattern, longer_alt: [exports.pNameLn] });
/**
 * [[142]](https://www.w3.org/TR/sparql11-query/#rBLANK_NODE_LABEL)
 */
exports.blankNodeLabel = (0, core_1.createToken)({ name: 'BlankNodeLabel', pattern: lexerPatterns_js_1.blankNodeLabelPattern });
/**
 * [[143]](https://www.w3.org/TR/sparql11-query/#rVAR1)
 */
exports.var1 = (0, core_1.createToken)({ name: 'Var1', pattern: lexerPatterns_js_1.var1Pattern });
/**
 * [[144]](https://www.w3.org/TR/sparql11-query/#rVAR1)
 */
exports.var2 = (0, core_1.createToken)({ name: 'Var2', pattern: lexerPatterns_js_1.var2Pattern });
/**
 * [[145]](https://www.w3.org/TR/sparql11-query/#rLANGTAG)
 */
exports.langTag = (0, core_1.createToken)({ name: 'LangTag', pattern: lexerPatterns_js_1.langTagPattern });
/**
 * [[146]](https://www.w3.org/TR/sparql11-query/#rINTEGER)
 */
exports.integer = (0, core_1.createToken)({ name: 'Integer', pattern: lexerPatterns_js_1.integerPattern });
/**
 * [[148]](https://www.w3.org/TR/sparql11-query/#rDOUBLE)
 */
exports.decimal = (0, core_1.createToken)({ name: 'Decimal', pattern: lexerPatterns_js_1.decimalPattern });
/**
 * [[148]](https://www.w3.org/TR/sparql11-query/#rDOUBLE)
 */
exports.double = (0, core_1.createToken)({ name: 'Double', pattern: lexerPatterns_js_1.doublePattern });
/**
 * [[149]](https://www.w3.org/TR/sparql11-query/#rINTEGER_POSITIVE)
 */
exports.integerPositive = (0, core_1.createToken)({ name: 'IntegerPositive', pattern: lexerPatterns_js_1.integerPositivePattern });
/**
 * [[150]](https://www.w3.org/TR/sparql11-query/#rDECIMAL_POSITIVE)
 */
exports.decimalPositive = (0, core_1.createToken)({ name: 'DecimalPositive', pattern: lexerPatterns_js_1.decimalPositivePattern });
/**
 * [[151]](https://www.w3.org/TR/sparql11-query/#rDOUBLE_POSITIVE)
 */
exports.doublePositive = (0, core_1.createToken)({ name: 'DoublePositive', pattern: lexerPatterns_js_1.doublePositivePattern });
/**
 * [[152]](https://www.w3.org/TR/sparql11-query/#rINTEGER_NEGATIVE)
 */
exports.integerNegative = (0, core_1.createToken)({ name: 'IntegerNegative', pattern: lexerPatterns_js_1.integerNegativePattern });
/**
 * [[153]](https://www.w3.org/TR/sparql11-query/#rDECIMAL_NEGATIVE)
 */
exports.decimalNegative = (0, core_1.createToken)({ name: 'DecimalNegative', pattern: lexerPatterns_js_1.decimalNegativePattern });
/**
 * [[154]](https://www.w3.org/TR/sparql11-query/#rDOUBLE_NEGATIVE)
 */
exports.doubleNegative = (0, core_1.createToken)({ name: 'DoubleNegative', pattern: lexerPatterns_js_1.doubleNegativePattern });
/**
 * [[156]](https://www.w3.org/TR/sparql11-query/#rSTRING_LITERAL1)
 */
exports.stringLiteral1 = (0, core_1.createToken)({ name: 'StringLiteral1', pattern: lexerPatterns_js_1.stringLiteral1Pattern });
/**
 * [[157]](https://www.w3.org/TR/sparql11-query/#rSTRING_LITERAL1)
 */
exports.stringLiteral2 = (0, core_1.createToken)({ name: 'StringLiteral2', pattern: lexerPatterns_js_1.stringLiteral2Pattern });
/**
 * [[158]](https://www.w3.org/TR/sparql11-query/#rSTRING_LITERAL_LONG1)
 */
exports.stringLiteralLong1 = (0, core_1.createToken)({ name: 'StringLiteralLong1', pattern: lexerPatterns_js_1.stringLiteralLong1Pattern });
/**
 * [[159]]https://www.w3.org/TR/sparql11-query/#rSTRING_LITERAL_LONG2)
 */
exports.stringLiteralLong2 = (0, core_1.createToken)({ name: 'StringLiteralLong2', pattern: lexerPatterns_js_1.stringLiteralLong2Pattern });
/**
 * https://www.w3.org/TR/sparql11-query/#rWS
 */
exports.ws = (0, core_1.createToken)({ name: 'Ws', pattern: lexerPatterns_js_1.wsPattern, group: chevrotain_1.Lexer.SKIPPED });
exports.comment = (0, core_1.createToken)({ name: 'Comment', pattern: lexerPatterns_js_1.commentPattern, group: chevrotain_1.Lexer.SKIPPED });
/**
 * [[161]](https://www.w3.org/TR/sparql11-query/#rNIL)
 */
exports.nil = (0, core_1.createToken)({ name: 'Nil', pattern: lexerPatterns_js_1.nilPattern });
/**
 * [[163]](https://www.w3.org/TR/sparql11-query/#rANON)
 */
exports.anon = (0, core_1.createToken)({ name: 'Anon', pattern: lexerPatterns_js_1.anonPattern });
exports.allTerminals = core_1.LexerBuilder.create().add(exports.iriRef, exports.pNameNs, exports.pNameLn, exports.blankNodeLabel, exports.var1, exports.var2, exports.langTag, exports.double, exports.decimal, exports.integer, exports.doublePositive, exports.decimalPositive, exports.integerPositive, exports.doubleNegative, exports.decimalNegative, exports.integerNegative, exports.stringLiteralLong1, exports.stringLiteralLong2, exports.stringLiteral1, exports.stringLiteral2, exports.ws, exports.comment, exports.nil, exports.anon);
