// @traqula/rules-sparql-1-2/dist/esm/lib/AstFactory.js
import { AstFactory as Sparql11Factory } from '@traqula/rules-sparql-1-1';
/**
 * A factory that helps you create, and check types for AST nodes for SPARQL 1.2
 */
export class AstFactory extends Sparql11Factory {
    termTriple(subject, predicate, object, loc) {
        return {
            type: 'term',
            subType: 'triple',
            subject,
            predicate,
            object,
            loc,
        };
    }
    isTermTriple(obj) {
        return this.isOfSubType(obj, 'term', 'triple');
    }
    tripleCollectionReifiedTriple(loc, subject, predicate, object, reifier) {
        return {
            type: 'tripleCollection',
            subType: 'reifiedTriple',
            triples: [this.triple(subject, predicate, object)],
            identifier: reifier ?? this.termBlank(undefined, this.sourceLocation()),
            loc,
        };
    }
    isTripleCollectionReifiedTriple(obj) {
        return this.isOfSubType(obj, 'tripleCollection', 'reifiedTriple');
    }
    tripleCollectionBlankNodeProperties(identifier, triples, loc) {
        return {
            type: 'tripleCollection',
            subType: 'blankNodeProperties',
            triples,
            identifier,
            loc,
        };
    }
    /**
     * Overwritten triple constructor to always contain an empty annotations list
     */
    triple = (subject, predicate, object, loc) => ({
        type: 'triple',
        subject,
        predicate,
        object,
        annotations: [],
        loc: loc ?? this.sourceLocation(subject, predicate, object),
    });
    annotatedTriple(subject, predicate, object, annotations, loc) {
        return {
            type: 'triple',
            subject,
            predicate,
            object,
            annotations: annotations ?? [],
            loc: loc ?? this.sourceLocation(subject, predicate, object, ...annotations ?? []),
        };
    }
    contextDefinitionVersion(version, loc) {
        return {
            type: 'contextDef',
            subType: 'version',
            version,
            loc,
        };
    }
    isContextDefinitionVersion(obj) {
        return this.isOfSubType(obj, 'contextDef', 'version');
    }
}
