// @traqula/rules-sparql-1-2/dist/esm/lib/parserUtils.js
import { TransformerSubTyped, traqulaIndentation, traqulaNewlineAlternative } from '@traqula/core';
import { AstFactory } from './AstFactory.js';
export function completeParseContext(context) {
    return {
        astFactory: context.astFactory ?? new AstFactory({ tracksSourceLocation: false }),
        baseIRI: context.baseIRI,
        prefixes: { ...context.prefixes },
        parseMode: context.parseMode ? new Set(context.parseMode) : new Set(['canParseVars', 'canCreateBlankNodes']),
        skipValidation: context.skipValidation ?? false,
    };
}
export function completeGeneratorContext(context) {
    return {
        astFactory: context.astFactory ?? new AstFactory(),
        origSource: context.origSource ?? '',
        offset: context.offset,
        indentInc: context.indentInc ?? 2,
        [traqulaIndentation]: context[traqulaIndentation] ?? 0,
        [traqulaNewlineAlternative]: context[traqulaNewlineAlternative] ?? ' ',
    };
}
export function copyParseContext(context) {
    return {
        ...context,
        prefixes: { ...context.prefixes },
        parseMode: new Set(context.parseMode),
    };
}
export class AstTransformer extends TransformerSubTyped {
}
