// @traqula/algebra-transformations-1-1/dist/esm/lib/toAst/core.js
import { AstFactory, AstTransformer } from '@traqula/rules-sparql-1-1';
import * as Algebra from '../algebra.js';
import { AlgebraFactory } from '../algebraFactory.js';
import { types } from '../toAlgebra/index.js';
export function createAstContext() {
    return {
        project: false,
        extend: [],
        group: [],
        aggregates: [],
        order: [],
        algebraFactory: new AlgebraFactory(),
        astFactory: new AstFactory(),
        transformer: new AstTransformer(),
    };
}
export const eTypes = Algebra.ExpressionTypes;
export const resetContext = {
    name: 'resetContext',
    fun: () => (c) => {
        c.project = false;
        c.extend = [];
        c.group = [];
        c.aggregates = [];
        c.order = [];
    },
};
export const registerProjection = {
    name: 'registerProjection',
    fun: () => (c, op) => {
        // GRAPH was added because the way graphs get added back here is not the same as how they get added in the future
        // ^ seems fine but might have to be changed if problems get detected in the future
        if (op.type !== types.EXTEND && op.type !== types.ORDER_BY && op.type !== types.GRAPH) {
            c.project = false;
        }
    },
};
