// @traqula/rules-sparql-1-1/dist/esm/lib/factoryMixins/GraphRefFactory.js
const nodeType = 'graphRef';
// eslint-disable-next-line ts/explicit-function-return-type
export function GraphRefFactoryMixin(Base) {
    return class GraphRefFactory extends Base {
        isGraphRef(obj) {
            return this.isOfType(obj, nodeType);
        }
        graphRefDefault(loc) {
            return {
                type: nodeType,
                subType: 'default',
                loc,
            };
        }
        isGraphRefDefault(graphRef) {
            return this.isOfSubType(graphRef, nodeType, 'default');
        }
        graphRefNamed(loc) {
            return {
                type: nodeType,
                subType: 'named',
                loc,
            };
        }
        isGraphRefNamed(graphRef) {
            return this.isOfSubType(graphRef, nodeType, 'named');
        }
        graphRefAll(loc) {
            return {
                type: nodeType,
                subType: 'all',
                loc,
            };
        }
        isGraphRefAll(graphRef) {
            return this.isOfSubType(graphRef, nodeType, 'all');
        }
        graphRefSpecific(graph, loc) {
            return {
                type: nodeType,
                subType: 'specific',
                graph,
                loc,
            };
        }
        isGraphRefSpecific(graphRef) {
            return this.isOfSubType(graphRef, nodeType, 'specific');
        }
    };
}
