// @traqula/rules-sparql-1-1/dist/esm/lib/factoryMixins/Patternfactory.js
const nodeType = 'pattern';
// eslint-disable-next-line ts/explicit-function-return-type
export function PatternFactoryMixin(Base) {
    return class PatternFactory extends Base {
        isPattern(obj) {
            return this.isOfType(obj, nodeType);
        }
        patternBgp(triples, loc) {
            return { type: nodeType, subType: 'bgp', triples, loc };
        }
        isPatternBgp(obj) {
            return this.isOfSubType(obj, nodeType, 'bgp');
        }
        patternGroup(patterns, loc) {
            return { type: nodeType, subType: 'group', patterns, loc };
        }
        isPatternGroup(obj) {
            return this.isOfSubType(obj, nodeType, 'group');
        }
        patternGraph(name, patterns, loc) {
            return { type: nodeType, subType: 'graph', name, patterns, loc };
        }
        isPatternGraph(obj) {
            return this.isOfSubType(obj, nodeType, 'graph');
        }
        patternOptional(patterns, loc) {
            return { type: nodeType, subType: 'optional', patterns, loc };
        }
        isPatternOptional(obj) {
            return this.isOfSubType(obj, nodeType, 'optional');
        }
        patternValues(variables, values, loc) {
            return { type: nodeType, subType: 'values', variables, values, loc };
        }
        isPatternValues(obj) {
            return this.isOfSubType(obj, nodeType, 'values');
        }
        patternFilter(expression, loc) {
            return {
                type: nodeType,
                subType: 'filter',
                expression,
                loc,
            };
        }
        isPatternFilter(obj) {
            return this.isOfSubType(obj, nodeType, 'filter');
        }
        patternBind(expression, variable, loc) {
            return {
                type: nodeType,
                subType: 'bind',
                expression,
                variable,
                loc,
            };
        }
        isPatternBind(obj) {
            return this.isOfSubType(obj, nodeType, 'bind');
        }
        patternUnion(patterns, loc) {
            return {
                type: nodeType,
                subType: 'union',
                patterns,
                loc,
            };
        }
        isPatternUnion(obj) {
            return this.isOfSubType(obj, nodeType, 'union');
        }
        patternMinus(patterns, loc) {
            return {
                type: nodeType,
                subType: 'minus',
                patterns,
                loc,
            };
        }
        isPatternMinus(obj) {
            return this.isOfSubType(obj, nodeType, 'minus');
        }
        patternService(name, patterns, silent, loc) {
            return {
                type: nodeType,
                subType: 'service',
                silent,
                name,
                patterns,
                loc,
            };
        }
        isPatternService(obj) {
            return this.isOfSubType(obj, nodeType, 'service');
        }
    };
}
