// @traqula/rules-sparql-1-1/dist/esm/lib/factoryMixins/QueryFactory.js
const nodeType = 'query';
// eslint-disable-next-line ts/explicit-function-return-type
export function QueryFactoryMixin(Base) {
    return class QueryFactory extends Base {
        isQuery(obj) {
            return this.isOfType(obj, nodeType);
        }
        isQuerySelect(obj) {
            return this.isOfSubType(obj, nodeType, 'select');
        }
        queryConstruct(loc, context, template, where, solutionModifiers, datasets, values) {
            return {
                type: 'query',
                subType: 'construct',
                context,
                template,
                where,
                solutionModifiers,
                datasets,
                values,
                loc,
            };
        }
        isQueryConstruct(obj) {
            return this.isOfSubType(obj, nodeType, 'construct');
        }
        isQueryDescribe(obj) {
            return this.isOfSubType(obj, nodeType, 'describe');
        }
        isQueryAsk(obj) {
            return this.isOfSubType(obj, nodeType, 'ask');
        }
        querySelect(arg, loc) {
            return {
                type: nodeType,
                subType: 'select',
                ...arg,
                loc,
            };
        }
    };
}
