// @traqula/rules-sparql-1-1/dist/esm/lib/factoryMixins/SolutionModifiersFactory.js
const nodeType = 'solutionModifier';
// eslint-disable-next-line ts/explicit-function-return-type
export function SolutionModifiersFactoryMixin(Base) {
    return class SolutionModifiersFactory extends Base {
        isSolutionModifier(obj) {
            return this.isOfType(obj, nodeType);
        }
        solutionModifierHaving(having, loc) {
            return {
                type: nodeType,
                subType: 'having',
                having,
                loc,
            };
        }
        isSolutionModifierHaving(obj) {
            return this.isOfSubType(obj, nodeType, 'having');
        }
        solutionModifierOrder(orderDefs, loc) {
            return {
                type: nodeType,
                subType: 'order',
                orderDefs,
                loc,
            };
        }
        isSolutionModifierOrder(obj) {
            return this.isOfSubType(obj, nodeType, 'order');
        }
        solutionModifierLimitOffset(limit, offset, loc) {
            return {
                type: nodeType,
                subType: 'limitOffset',
                limit,
                offset,
                loc,
            };
        }
        isSolutionModifierLimitOffset(obj) {
            return this.isOfSubType(obj, nodeType, 'limitOffset');
        }
        solutionModifierGroup(groupings, loc) {
            return {
                type: 'solutionModifier',
                subType: 'group',
                groupings,
                loc,
            };
        }
        isSolutionModifierGroup(obj) {
            return this.isOfSubType(obj, nodeType, 'group');
        }
    };
}
