// @traqula/core/dist/esm/lib/transformers/TransformerObject.js
export class TransformerObject {
    defaultContext;
    maxStackSize = 1_000_000;
    /**
     * Creates stateless transformer.
     * @param defaultContext
     */
    constructor(defaultContext = {}) {
        this.defaultContext = defaultContext;
    }
    clone(newDefaultContext = {}) {
        return new TransformerObject({ ...this.defaultContext, ...newDefaultContext });
    }
    /**
     * Function to shallow clone any type.
     * @param obj
     * @protected
     */
    cloneObj(obj) {
        if (obj === null || typeof obj !== 'object') {
            return obj;
        }
        const proto = Object.getPrototypeOf(obj);
        // Fast path: plain object
        if (proto === Object.prototype || proto === null) {
            // Spread or assign preserves fast properties
            return { ...obj };
        }
        // Otherwise, preserve prototype for custom objects
        return Object.assign(Object.create(proto), obj);
    }
    /**
     * Recursively transforms all objects that are not arrays. Mapper is called on deeper objects first.
     * @param startObject object to start iterating from
     * @param mapper mapper to transform the various objects - argument is a copy of the original
     * @param preVisitor callback that is evaluated before iterating deeper.
     *   If continues is false, we do not iterate deeper, current object is still mapped. - default: true
     *   If shortcut is true, we do not iterate deeper, nor do we branch out, this mapper will be the last one called.
     *    - Default false
     */
    transformObject(startObject, mapper, preVisitor = () => ({})) {
        const defaults = this.defaultContext;
        const defaultCopyFlag = defaults.copy ?? true;
        const defaultContinues = defaults.continue ?? true;
        const defaultIgnoreKeys = defaults.ignoreKeys;
        const defaultShallowKeys = defaults.shallowKeys;
        const defaultDidShortCut = defaults.shortcut ?? false;
        // Code handles own stack instead of using recursion - this optimizes it for deep operations.
        let didShortCut = false;
        const resultWrap = { res: startObject };
        // Grows with stack
        const stack = [startObject];
        const stackParent = [resultWrap];
        const stackParentKey = ['res'];
        // Grows with reverse stack - when popping down the stack, you realise you still want to map something.
        // Counter of stack size when we started adding the children of this object, going beyond this means a new parent
        const handleMapperOnLen = [];
        const mapperCopyStack = [];
        const mapperOrigStack = [];
        const mapperParent = [];
        const mapperParentKey = [];
        function handleMapper() {
            while (stack.length === handleMapperOnLen.at(-1)) {
                handleMapperOnLen.pop();
                const copyToMap = mapperCopyStack.pop();
                const origToMap = mapperOrigStack.pop();
                const parent = mapperParent.pop();
                const parentKey = mapperParentKey.pop();
                parent[parentKey] = mapper(copyToMap, origToMap);
            }
        }
        while (stack.length > 0 && stack.length < this.maxStackSize) {
            const curObject = stack.pop();
            const curParent = stackParent.pop();
            const curKey = stackParentKey.pop();
            // Only add to the stack when you did not shortcut
            if (!didShortCut) {
                if (Array.isArray(curObject)) {
                    const newArr = [...curObject];
                    handleMapperOnLen.push(stack.length);
                    mapperCopyStack.push(newArr);
                    mapperOrigStack.push(curObject);
                    mapperParent.push(curParent);
                    mapperParentKey.push(curKey);
                    for (let index = curObject.length - 1; index >= 0; index--) {
                        const val = curObject[index];
                        if (val !== null && typeof val === 'object') {
                            stack.push(val);
                            stackParent.push(newArr);
                            stackParentKey.push(index.toString());
                        }
                    }
                    handleMapper();
                    continue;
                }
                // Perform pre visit before expanding the stack
                const context = preVisitor(curObject);
                const copyFlag = context.copy ?? defaultCopyFlag;
                const continues = context.continue ?? defaultContinues;
                const ignoreKeys = context.ignoreKeys ?? defaultIgnoreKeys;
                const shallowKeys = context.shallowKeys ?? defaultShallowKeys;
                didShortCut = context.shortcut ?? defaultDidShortCut;
                const copy = copyFlag ? this.cloneObj(curObject) : curObject;
                // Register that you want to be visited
                handleMapperOnLen.push(stack.length);
                mapperCopyStack.push(copy);
                mapperOrigStack.push(curObject);
                mapperParent.push(curParent);
                mapperParentKey.push(curKey);
                // Extend stack if needed. When shortcutted, should still unwind the stack, but no longer add to it.
                if (continues && !didShortCut) {
                    for (const key in copy) {
                        if (!Object.hasOwn(copy, key)) {
                            continue;
                        }
                        const val = copy[key];
                        // If shallow copy required, do
                        const onlyShallow = shallowKeys && shallowKeys?.has(key);
                        if (onlyShallow) {
                            // Do not add stack entry - assign straight away
                            copy[key] = this.cloneObj(val);
                        }
                        if (ignoreKeys && ignoreKeys.has(key)) {
                            // Do not add stack entry
                            continue;
                        }
                        if (!onlyShallow && val !== null && typeof val === 'object') {
                            // Do add stack entry.
                            stack.push(val);
                            stackParentKey.push(key);
                            stackParent.push(copy);
                        }
                    }
                }
            }
            handleMapper();
        }
        if (stack.length >= this.maxStackSize) {
            throw new Error('Transform object stack overflowed');
        }
        handleMapper();
        return resultWrap.res;
    }
    /**
     * Visitor that visits all objects. Visits deeper objects first.
     */
    visitObject(startObject, visitor, preVisitor = () => ({})) {
        const defaults = this.defaultContext;
        const defaultContinues = defaults.continue ?? true;
        const defaultIgnoreKeys = defaults.ignoreKeys;
        const defaultShortcut = defaults.shortcut ?? false;
        let didShortCut = false;
        // Stack of things to preVisit
        const stack = [startObject];
        // When the stack is done preVisiting things above this lengths, visit the bellow
        const handleVisitorOnLen = [];
        const visitorStack = [];
        function handleVisitor() {
            while (stack.length === handleVisitorOnLen.at(-1)) {
                handleVisitorOnLen.pop();
                const toVisit = visitorStack.pop();
                visitor(toVisit);
            }
        }
        while (stack.length > 0 && stack.length < this.maxStackSize) {
            const curObject = stack.pop();
            if (!didShortCut) {
                if (Array.isArray(curObject)) {
                    for (let i = curObject.length - 1; i >= 0; i--) {
                        stack.push(curObject[i]);
                    }
                    handleVisitor();
                    continue;
                }
                // Perform pre visit before expanding the stack
                const context = preVisitor(curObject);
                didShortCut = context.shortcut ?? defaultShortcut;
                const continues = context.continue ?? defaultContinues;
                const ignoreKeys = context.ignoreKeys ?? defaultIgnoreKeys;
                // Register that you want to be visited
                handleVisitorOnLen.push(stack.length);
                visitorStack.push(curObject);
                // Extend stack if needed. When shortcutted, should still unwind the stack, but no longer add to it.
                if (continues && !didShortCut) {
                    for (const key in curObject) {
                        if (!Object.hasOwn(curObject, key)) {
                            continue;
                        }
                        if (ignoreKeys && ignoreKeys.has(key)) {
                            continue;
                        }
                        const val = curObject[key];
                        if (val && typeof val === 'object') {
                            stack.push(val);
                        }
                    }
                }
            }
            handleVisitor();
        }
        if (stack.length >= this.maxStackSize) {
            throw new Error('Transform object stack overflowed');
        }
        handleVisitor();
    }
}
