// @traqula/algebra-transformations-1-1/dist/cjs/lib/toAst/core.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerProjection = exports.resetContext = exports.eTypes = void 0;
exports.createAstContext = createAstContext;
const rules_sparql_1_1_1 = require("@traqula/rules-sparql-1-1");
const Algebra = __importStar(require("../algebra.js"));
const algebraFactory_js_1 = require("../algebraFactory.js");
const index_js_1 = require("../toAlgebra/index.js");
function createAstContext() {
    return {
        project: false,
        extend: [],
        group: [],
        aggregates: [],
        order: [],
        algebraFactory: new algebraFactory_js_1.AlgebraFactory(),
        astFactory: new rules_sparql_1_1_1.AstFactory(),
        transformer: new rules_sparql_1_1_1.AstTransformer(),
    };
}
exports.eTypes = Algebra.ExpressionTypes;
exports.resetContext = {
    name: 'resetContext',
    fun: () => (c) => {
        c.project = false;
        c.extend = [];
        c.group = [];
        c.aggregates = [];
        c.order = [];
    },
};
exports.registerProjection = {
    name: 'registerProjection',
    fun: () => (c, op) => {
        // GRAPH was added because the way graphs get added back here is not the same as how they get added in the future
        // ^ seems fine but might have to be changed if problems get detected in the future
        if (op.type !== index_js_1.types.EXTEND && op.type !== index_js_1.types.ORDER_BY && op.type !== index_js_1.types.GRAPH) {
            c.project = false;
        }
    },
};
