// @traqula/algebra-transformations-1-1/dist/cjs/lib/toAst/path.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.translateAlgZeroOrOnePath = exports.translateAlgZeroOrMorePath = exports.translateAlgSeq = exports.translateAlgOneOrMorePath = exports.translateAlgNps = exports.translateAlgLink = exports.translateAlgInv = exports.translateAlgAlt = exports.translateAlgPathComponent = void 0;
const core_js_1 = require("../toAlgebra/core.js");
const general_js_1 = require("./general.js");
exports.translateAlgPathComponent = {
    name: 'translatePathComponent',
    fun: ({ SUBRULE }) => (_, path) => {
        switch (path.type) {
            case core_js_1.types.ALT: return SUBRULE(exports.translateAlgAlt, path);
            case core_js_1.types.INV: return SUBRULE(exports.translateAlgInv, path);
            case core_js_1.types.LINK: return SUBRULE(exports.translateAlgLink, path);
            case core_js_1.types.NPS: return SUBRULE(exports.translateAlgNps, path);
            case core_js_1.types.ONE_OR_MORE_PATH: return SUBRULE(exports.translateAlgOneOrMorePath, path);
            case core_js_1.types.SEQ: return SUBRULE(exports.translateAlgSeq, path);
            case core_js_1.types.ZERO_OR_MORE_PATH: return SUBRULE(exports.translateAlgZeroOrMorePath, path);
            case core_js_1.types.ZERO_OR_ONE_PATH: return SUBRULE(exports.translateAlgZeroOrOnePath, path);
            default:
                throw new Error(`Unknown Path type ${path.type}`);
        }
    },
};
exports.translateAlgAlt = {
    name: 'translateAlt',
    fun: ({ SUBRULE }) => ({ astFactory: F }, path) => {
        const mapped = path.input.map(x => SUBRULE(exports.translateAlgPathComponent, x));
        if (mapped.every(entry => F.isPathOfType(entry, ['!']))) {
            return F.path('!', [F.path('|', mapped.flatMap(entry => entry.items), F.gen())], F.gen());
        }
        return F.path('|', mapped, F.gen());
    },
};
exports.translateAlgInv = {
    name: 'translateInv',
    fun: ({ SUBRULE }) => ({ astFactory: F }, path) => {
        if (path.path.type === core_js_1.types.NPS) {
            const inv = path.path.iris.map((iri) => F.path('^', [SUBRULE(general_js_1.translateAlgTerm, iri)], F.gen()));
            if (inv.length <= 1) {
                return F.path('!', inv, F.gen());
            }
            return F.path('!', [F.path('|', inv, F.gen())], F.gen());
        }
        return F.path('^', [SUBRULE(exports.translateAlgPathComponent, path.path)], F.gen());
    },
};
exports.translateAlgLink = {
    name: 'translateLink',
    fun: ({ SUBRULE }) => (_, path) => SUBRULE(general_js_1.translateAlgTerm, path.iri),
};
exports.translateAlgNps = {
    name: 'translateNps',
    fun: ({ SUBRULE }) => ({ astFactory: F }, path) => {
        if (path.iris.length === 1) {
            return F.path('!', [SUBRULE(general_js_1.translateAlgTerm, path.iris[0])], F.gen());
        }
        return F.path('!', [
            F.path('|', path.iris.map(x => SUBRULE(general_js_1.translateAlgTerm, x)), F.gen()),
        ], F.gen());
    },
};
exports.translateAlgOneOrMorePath = {
    name: 'translateOneOrMorePath',
    fun: ({ SUBRULE }) => ({ astFactory: F }, path) => F.path('+', [SUBRULE(exports.translateAlgPathComponent, path.path)], F.gen()),
};
exports.translateAlgSeq = {
    name: 'translateSeq',
    fun: ({ SUBRULE }) => ({ astFactory: F }, path) => F.path('/', path.input.map(x => SUBRULE(exports.translateAlgPathComponent, x)), F.gen()),
};
exports.translateAlgZeroOrMorePath = {
    name: 'translateZeroOrMorePath',
    fun: ({ SUBRULE }) => ({ astFactory: F }, path) => F.path('*', [SUBRULE(exports.translateAlgPathComponent, path.path)], F.gen()),
};
exports.translateAlgZeroOrOnePath = {
    name: 'translateZeroOrOnePath',
    fun: ({ SUBRULE }) => ({ astFactory: F }, path) => F.path('?', [SUBRULE(exports.translateAlgPathComponent, path.path)], F.gen()),
};
