// @traqula/algebra-transformations-1-1/dist/cjs/lib/algebra.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExpressionTypes = exports.Types = void 0;
var Types;
(function (Types) {
    Types["ASK"] = "ask";
    Types["BGP"] = "bgp";
    Types["CONSTRUCT"] = "construct";
    Types["DESCRIBE"] = "describe";
    Types["DISTINCT"] = "distinct";
    Types["EXPRESSION"] = "expression";
    Types["EXTEND"] = "extend";
    Types["FILTER"] = "filter";
    Types["FROM"] = "from";
    Types["GRAPH"] = "graph";
    Types["GROUP"] = "group";
    Types["JOIN"] = "join";
    Types["LEFT_JOIN"] = "leftjoin";
    Types["MINUS"] = "minus";
    Types["NOP"] = "nop";
    Types["ORDER_BY"] = "orderby";
    Types["PATTERN"] = "pattern";
    Types["PROJECT"] = "project";
    Types["REDUCED"] = "reduced";
    Types["SERVICE"] = "service";
    Types["SLICE"] = "slice";
    Types["UNION"] = "union";
    Types["VALUES"] = "values";
    // Update
    Types["COMPOSITE_UPDATE"] = "compositeupdate";
    Types["DELETE_INSERT"] = "deleteinsert";
    Types["LOAD"] = "load";
    Types["CLEAR"] = "clear";
    Types["CREATE"] = "create";
    Types["DROP"] = "drop";
    Types["ADD"] = "add";
    Types["MOVE"] = "move";
    Types["COPY"] = "copy";
    // Paths
    Types["PATH"] = "path";
    Types["ALT"] = "alt";
    Types["INV"] = "inv";
    Types["LINK"] = "link";
    Types["ONE_OR_MORE_PATH"] = "OneOrMorePath";
    Types["SEQ"] = "seq";
    Types["NPS"] = "nps";
    Types["ZERO_OR_MORE_PATH"] = "ZeroOrMorePath";
    Types["ZERO_OR_ONE_PATH"] = "ZeroOrOnePath";
})(Types || (exports.Types = Types = {}));
var ExpressionTypes;
(function (ExpressionTypes) {
    ExpressionTypes["AGGREGATE"] = "aggregate";
    ExpressionTypes["EXISTENCE"] = "existence";
    ExpressionTypes["NAMED"] = "named";
    ExpressionTypes["OPERATOR"] = "operator";
    ExpressionTypes["TERM"] = "term";
    ExpressionTypes["WILDCARD"] = "wildcard";
})(ExpressionTypes || (exports.ExpressionTypes = ExpressionTypes = {}));
