// @traqula/algebra-transformations-1-1/dist/cjs/lib/canonicalizer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Canonicalizer = void 0;
const rdf_data_factory_1 = require("rdf-data-factory");
const Algebra = __importStar(require("./algebra.js"));
const util = __importStar(require("./util.js"));
const index_js_1 = require("./index.js");
class Canonicalizer {
    constructor() {
        this.blankId = 0;
    }
    blankId;
    genValue() {
        return `value_${this.blankId++}`;
    }
    /**
     * Replaces values of BlankNodes in a query with newly generated names.
     * @param res
     * @param replaceVariables
     */
    canonicalizeQuery(res, replaceVariables) {
        this.blankId = 0;
        const nameMapping = {};
        const factory = new index_js_1.AlgebraFactory();
        return util.mapOperation(res, {
            [Algebra.Types.PATH]: { transform: pathOp => ({
                    result: factory.createPath(this.replaceValue(pathOp.subject, nameMapping, replaceVariables, factory), pathOp.predicate, this.replaceValue(pathOp.object, nameMapping, replaceVariables, factory), this.replaceValue(pathOp.graph, nameMapping, replaceVariables, factory)),
                    recurse: true,
                }) },
            [Algebra.Types.PATTERN]: { transform: patternOp => ({
                    result: factory.createPattern(this.replaceValue(patternOp.subject, nameMapping, replaceVariables, factory), this.replaceValue(patternOp.predicate, nameMapping, replaceVariables, factory), this.replaceValue(patternOp.object, nameMapping, replaceVariables, factory), this.replaceValue(patternOp.graph, nameMapping, replaceVariables, factory)),
                    recurse: true,
                }) },
            [Algebra.Types.CONSTRUCT]: { transform: constructOp => 
                // Blank nodes in CONSTRUCT templates must be maintained
                ({
                    result: factory.createConstruct(constructOp.input, constructOp.template),
                    recurse: true,
                }) },
        });
    }
    replaceValue(term, nameMapping, replaceVars, factory) {
        if (term.termType === 'Quad') {
            return factory.createPattern(this.replaceValue(term.subject, nameMapping, replaceVars, factory), this.replaceValue(term.predicate, nameMapping, replaceVars, factory), this.replaceValue(term.object, nameMapping, replaceVars, factory), this.replaceValue(term.graph, nameMapping, replaceVars, factory));
        }
        if (term.termType !== 'BlankNode' && (term.termType !== 'Variable' || !replaceVars)) {
            return term;
        }
        const dataFactory = new rdf_data_factory_1.DataFactory();
        const generateTerm = term.termType === 'Variable' ?
            dataFactory.variable.bind(dataFactory) :
            dataFactory.blankNode.bind(dataFactory);
        let val = nameMapping[term.value];
        if (!val) {
            val = this.genValue();
            nameMapping[term.value] = val;
        }
        return generateTerm(val);
    }
}
exports.Canonicalizer = Canonicalizer;
