// @traqula/algebra-transformations-1-1/dist/esm/lib/toAlgebra/core.js
import { AstFactory, AstTransformer } from '@traqula/rules-sparql-1-1';
import { DataFactory } from 'rdf-data-factory';
import * as Algebra from '../algebra.js';
import { AlgebraFactory } from '../algebraFactory.js';
export function createAlgebraContext(config) {
    const dataFactory = config.dataFactory ?? new DataFactory();
    return {
        variables: new Set(),
        varCount: 0,
        useQuads: false,
        transformer: new AstTransformer(),
        astFactory: new AstFactory(),
        dataFactory,
        algebraFactory: new AlgebraFactory(dataFactory),
        currentBase: config.baseIRI,
        currentPrefixes: config.prefixes ? { ...config.prefixes } : {},
    };
}
export const types = Algebra.Types;
export const typeVals = Object.values(types);
export function isTerm(term) {
    return Boolean(term?.termType);
}
// This is not completely correct but this way we also catch SPARQL.js triples
export function isTriple(triple) {
    return triple.subject && triple.predicate && triple.object;
}
export function isVariable(term) {
    return term?.termType === 'Variable';
}
