// @traqula/rules-sparql-1-2/dist/cjs/lib/lexer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sparql12LexerBuilder = exports.LANG_DIR = exports.buildInOBJECT = exports.buildInPREDICATE = exports.buildInSUBJECT = exports.buildInTRIPLE = exports.buildInIsTRIPLE = exports.buildInHasLangDir = exports.buildInHasLang = exports.buildInStrLangDir = exports.buildInLangDir = exports.tripleTermClose = exports.tripleTermOpen = exports.reificationClose = exports.reificationOpen = exports.annotationClose = exports.annotationOpen = exports.tilde = exports.version = void 0;
/* eslint-disable require-unicode-regexp */
const core_1 = require("@traqula/core");
const rules_sparql_1_1_1 = require("@traqula/rules-sparql-1-1");
exports.version = (0, core_1.createToken)({ name: 'Version', pattern: /version/i, label: 'version identifier' });
exports.tilde = (0, core_1.createToken)({ name: 'Tilde', pattern: '~', label: '~' });
exports.annotationOpen = (0, core_1.createToken)({ name: 'AnnotationOpen', pattern: '{|', label: `Annotation Open: {|` });
exports.annotationClose = (0, core_1.createToken)({ name: 'AnnotationClose', pattern: '|}', label: 'Annotation Close |}' });
exports.reificationOpen = (0, core_1.createToken)({ name: 'ReificationOpen', pattern: '<<', label: 'Reification open <<' });
exports.reificationClose = (0, core_1.createToken)({ name: 'ReificationClose', pattern: '>>', label: 'Reification close >>' });
exports.tripleTermOpen = (0, core_1.createToken)({ name: 'TripleTermOpen', pattern: '<<(', label: 'Triple Term Open <<(' });
exports.tripleTermClose = (0, core_1.createToken)({ name: 'TripleTermClose', pattern: ')>>', label: 'Triple Term Close )>>' });
exports.buildInLangDir = (0, core_1.createToken)({ name: 'BuiltInLangdir', pattern: /langdir/i, label: 'LANGDIR' });
exports.buildInStrLangDir = (0, core_1.createToken)({
    name: 'BuiltInStrLangdir',
    pattern: /strlangdir/i,
    label: 'STRLANGDIR',
});
exports.buildInHasLang = (0, core_1.createToken)({ name: 'BuiltInHasLang', pattern: /haslang/i, label: 'hasLANG' });
exports.buildInHasLangDir = (0, core_1.createToken)({
    name: 'BuiltInHasLangdir',
    pattern: /haslangdir/i,
    label: 'hasLANGDIR',
});
exports.buildInIsTRIPLE = (0, core_1.createToken)({ name: 'BuiltInIsTriple', pattern: /istriple/i, label: 'isTRIPLE' });
exports.buildInTRIPLE = (0, core_1.createToken)({ name: 'BuiltInTriple', pattern: /triple/i, label: 'TRIPLE' });
exports.buildInSUBJECT = (0, core_1.createToken)({ name: 'BuiltInSubject', pattern: /subject/i, label: 'SUBJECT' });
exports.buildInPREDICATE = (0, core_1.createToken)({ name: 'BuiltInPredicate', pattern: /predicate/i, label: 'PREDICATE' });
exports.buildInOBJECT = (0, core_1.createToken)({ name: 'BuiltInObject', pattern: /object/i, label: 'OBJECT' });
exports.LANG_DIR = (0, core_1.createToken)({
    name: 'LANG_DIR',
    pattern: /@[a-z]+(?:-[\da-z]+)*(?:--[a-z]+)?/i,
    label: 'LANG_DIR',
});
exports.sparql12LexerBuilder = core_1.LexerBuilder
    .create(rules_sparql_1_1_1.lex.sparql11LexerBuilder)
    .addBefore(rules_sparql_1_1_1.lex.symbols.logicAnd, exports.tilde, exports.annotationOpen, exports.annotationClose, exports.tripleTermOpen, exports.tripleTermClose, exports.reificationOpen, exports.reificationClose, exports.version)
    .addBefore(rules_sparql_1_1_1.lex.builtIn.langmatches, exports.buildInLangDir, exports.buildInStrLangDir, exports.buildInHasLangDir, exports.buildInHasLang, exports.buildInIsTRIPLE, exports.buildInTRIPLE, exports.buildInSUBJECT, exports.buildInPREDICATE, exports.buildInOBJECT)
    .addBefore(rules_sparql_1_1_1.lex.terminals.langTag, exports.LANG_DIR)
    .delete(rules_sparql_1_1_1.lex.terminals.langTag);
