// @traqula/rules-sparql-1-2/dist/cjs/lib/parserUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AstTransformer = void 0;
exports.completeParseContext = completeParseContext;
exports.completeGeneratorContext = completeGeneratorContext;
exports.copyParseContext = copyParseContext;
const core_1 = require("@traqula/core");
const AstFactory_js_1 = require("./AstFactory.js");
function completeParseContext(context) {
    return {
        astFactory: context.astFactory ?? new AstFactory_js_1.AstFactory({ tracksSourceLocation: false }),
        baseIRI: context.baseIRI,
        prefixes: { ...context.prefixes },
        parseMode: context.parseMode ? new Set(context.parseMode) : new Set(['canParseVars', 'canCreateBlankNodes']),
        skipValidation: context.skipValidation ?? false,
    };
}
function completeGeneratorContext(context) {
    return {
        astFactory: context.astFactory ?? new AstFactory_js_1.AstFactory(),
        origSource: context.origSource ?? '',
        offset: context.offset,
        indentInc: context.indentInc ?? 2,
        [core_1.traqulaIndentation]: context[core_1.traqulaIndentation] ?? 0,
        [core_1.traqulaNewlineAlternative]: context[core_1.traqulaNewlineAlternative] ?? ' ',
    };
}
function copyParseContext(context) {
    return {
        ...context,
        prefixes: { ...context.prefixes },
        parseMode: new Set(context.parseMode),
    };
}
class AstTransformer extends core_1.TransformerSubTyped {
}
exports.AstTransformer = AstTransformer;
