// @traqula/rules-sparql-1-1/dist/cjs/lib/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AstTransformer = exports.CommonIRIs = void 0;
exports.sparqlCodepointEscape = sparqlCodepointEscape;
const core_1 = require("@traqula/core");
/**
 * Transform input in accordance to [19.2](https://www.w3.org/TR/sparql11-query/#codepointEscape)
 * and validate unicode codepoints.
 */
function sparqlCodepointEscape(input) {
    const sanitizedInput = input.replaceAll(/\\u([0-9a-fA-F]{4})|\\U([0-9a-fA-F]{8})/gu, (_, unicode4, unicode8) => {
        if (unicode4) {
            const charCode = Number.parseInt(unicode4, 16);
            return String.fromCodePoint(charCode);
        }
        const charCode = Number.parseInt(unicode8, 16);
        if (charCode < 0xFFFF) {
            return String.fromCodePoint(charCode);
        }
        const substractedCharCode = charCode - 0x10000;
        return String.fromCodePoint(0xD800 + (substractedCharCode >> 10), 0xDC00 + (substractedCharCode & 0x3FF));
    });
    // Test for invalid unicode surrogate pairs
    if (/[\uD800-\uDBFF](?:[^\uDC00-\uDFFF]|$)/u.test(sanitizedInput)) {
        throw new Error(`Invalid unicode codepoint of surrogate pair without corresponding codepoint`);
    }
    return sanitizedInput;
}
var CommonIRIs;
(function (CommonIRIs) {
    // XSD
    CommonIRIs["BOOLEAN"] = "http://www.w3.org/2001/XMLSchema#boolean";
    CommonIRIs["INTEGER"] = "http://www.w3.org/2001/XMLSchema#integer";
    CommonIRIs["DECIMAL"] = "http://www.w3.org/2001/XMLSchema#decimal";
    CommonIRIs["DOUBLE"] = "http://www.w3.org/2001/XMLSchema#double";
    CommonIRIs["STRING"] = "http://www.w3.org/2001/XMLSchema#string";
    // RDF
    CommonIRIs["FIRST"] = "http://www.w3.org/1999/02/22-rdf-syntax-ns#first";
    CommonIRIs["REST"] = "http://www.w3.org/1999/02/22-rdf-syntax-ns#rest";
    CommonIRIs["NIL"] = "http://www.w3.org/1999/02/22-rdf-syntax-ns#nil";
    CommonIRIs["TYPE"] = "http://www.w3.org/1999/02/22-rdf-syntax-ns#type";
})(CommonIRIs || (exports.CommonIRIs = CommonIRIs = {}));
class AstTransformer extends core_1.TransformerSubTyped {
}
exports.AstTransformer = AstTransformer;
