// @traqula/rules-sparql-1-1/dist/cjs/lib/factoryMixins/UpdateOperationFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpdateOperationFactoryMixin = UpdateOperationFactoryMixin;
const nodeType = 'updateOperation';
// eslint-disable-next-line ts/explicit-function-return-type
function UpdateOperationFactoryMixin(Base) {
    return class UpdateOperationFactory extends Base {
        isUpdateOperation(obj) {
            return this.isOfType(obj, nodeType);
        }
        updateOperationLoad(loc, source, silent, destination) {
            return {
                type: nodeType,
                subType: 'load',
                silent,
                source,
                ...(destination && { destination }),
                loc,
            };
        }
        isUpdateOperationLoad(obj) {
            return this.isOfSubType(obj, nodeType, 'load');
        }
        updateOperationClearDrop(subType, silent, destination, loc) {
            return {
                type: 'updateOperation',
                subType,
                silent,
                destination,
                loc,
            };
        }
        updateOperationClear(destination, silent, loc) {
            return this.updateOperationClearDrop('clear', silent, destination, loc);
        }
        ;
        isUpdateOperationClear(obj) {
            return this.isOfSubType(obj, nodeType, 'clear');
        }
        updateOperationDrop(destination, silent, loc) {
            return this.updateOperationClearDrop('drop', silent, destination, loc);
        }
        isUpdateOperationDrop(obj) {
            return this.isOfSubType(obj, nodeType, 'drop');
        }
        updateOperationCreate(destination, silent, loc) {
            return {
                type: 'updateOperation',
                subType: 'create',
                silent,
                destination,
                loc,
            };
        }
        isUpdateOperationCreate(obj) {
            return this.isOfSubType(obj, nodeType, 'create');
        }
        updateOperationAddMoveCopy(subType, source, destination, silent, loc) {
            return {
                type: 'updateOperation',
                subType,
                silent,
                source,
                destination,
                loc,
            };
        }
        updateOperationAdd(source, destination, silent, loc) {
            return this.updateOperationAddMoveCopy('add', source, destination, silent, loc);
        }
        isUpdateOperationAdd(obj) {
            return this.isOfSubType(obj, nodeType, 'add');
        }
        updateOperationMove(source, destination, silent, loc) {
            return this.updateOperationAddMoveCopy('move', source, destination, silent, loc);
        }
        isUpdateOperationMove(obj) {
            return this.isOfSubType(obj, nodeType, 'move');
        }
        updateOperationCopy(source, destination, silent, loc) {
            return this.updateOperationAddMoveCopy('copy', source, destination, silent, loc);
        }
        isUpdateOperationCopy(obj) {
            return this.isOfSubType(obj, nodeType, 'copy');
        }
        updateOperationInsDelDataWhere(subType, data, loc) {
            return {
                type: 'updateOperation',
                subType,
                data,
                loc,
            };
        }
        updateOperationInsertData(data, loc) {
            return this.updateOperationInsDelDataWhere('insertdata', data, loc);
        }
        isUpdateOperationInsertData(obj) {
            return this.isOfSubType(obj, nodeType, 'insertdata');
        }
        updateOperationDeleteData(data, loc) {
            return this.updateOperationInsDelDataWhere('deletedata', data, loc);
        }
        isUpdateOperationDeleteData(obj) {
            return this.isOfSubType(obj, nodeType, 'deletedata');
        }
        updateOperationDeleteWhere(data, loc) {
            return this.updateOperationInsDelDataWhere('deletewhere', data, loc);
        }
        isUpdateOperationDeleteWhere(obj) {
            return this.isOfSubType(obj, nodeType, 'deletewhere');
        }
        updateOperationModify(loc, insert, del, where, from, graph) {
            return {
                type: 'updateOperation',
                subType: 'modify',
                insert: insert ?? [],
                delete: del ?? [],
                graph,
                where,
                from,
                loc,
            };
        }
        isUpdateOperationModify(obj) {
            return this.isOfSubType(obj, nodeType, 'modify');
        }
    };
}
