// @traqula/algebra-transformations-1-1/dist/cjs/lib/toAlgebra/core.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.typeVals = exports.types = void 0;
exports.createAlgebraContext = createAlgebraContext;
exports.isTerm = isTerm;
exports.isTriple = isTriple;
exports.isVariable = isVariable;
const rules_sparql_1_1_1 = require("@traqula/rules-sparql-1-1");
const rdf_data_factory_1 = require("rdf-data-factory");
const Algebra = __importStar(require("../algebra.js"));
const algebraFactory_js_1 = require("../algebraFactory.js");
function createAlgebraContext(config) {
    const dataFactory = config.dataFactory ?? new rdf_data_factory_1.DataFactory();
    return {
        variables: new Set(),
        varCount: 0,
        useQuads: false,
        transformer: new rules_sparql_1_1_1.AstTransformer(),
        astFactory: new rules_sparql_1_1_1.AstFactory(),
        dataFactory,
        algebraFactory: new algebraFactory_js_1.AlgebraFactory(dataFactory),
        currentBase: config.baseIRI,
        currentPrefixes: config.prefixes ? { ...config.prefixes } : {},
    };
}
exports.types = Algebra.Types;
exports.typeVals = Object.values(exports.types);
function isTerm(term) {
    return Boolean(term?.termType);
}
// This is not completely correct but this way we also catch SPARQL.js triples
function isTriple(triple) {
    return triple.subject && triple.predicate && triple.object;
}
function isVariable(term) {
    return term?.termType === 'Variable';
}
