// @traqula/algebra-transformations-1-1/dist/cjs/lib/toAlgebra/patterns.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.simplifiedJoin = exports.accumulateGroupGraphPattern = exports.translateBgp = exports.translateGraphPattern = exports.translateExpression = void 0;
const core_js_1 = require("./core.js");
const general_js_1 = require("./general.js");
const path_js_1 = require("./path.js");
const toAlgebra_js_1 = require("./toAlgebra.js");
const tripleAndQuad_js_1 = require("./tripleAndQuad.js");
/**
 * Handles part of: 18.2.2.2 Collect FILTER Elements
 */
exports.translateExpression = {
    name: 'translateExpression',
    fun: ({ SUBRULE }) => ({ astFactory: F, algebraFactory: AF }, expr) => {
        if (F.isTerm(expr)) {
            return AF.createTermExpression(SUBRULE(general_js_1.translateTerm, expr));
        }
        if (F.isWildcard(expr)) {
            return AF.createWildcardExpression();
        }
        if (F.isExpressionAggregate(expr)) {
            return AF.createAggregateExpression(expr.aggregation, SUBRULE(exports.translateExpression, expr.expression[0]), expr.distinct, F.isExpressionAggregateSeparator(expr) ? expr.separator : undefined);
        }
        if (F.isExpressionFunctionCall(expr)) {
            // Outdated typings
            return AF.createNamedExpression(SUBRULE(general_js_1.translateNamed, expr.function), expr.args.map(subExpr => SUBRULE(exports.translateExpression, subExpr)));
        }
        if (F.isExpressionOperator(expr)) {
            return AF.createOperatorExpression(expr.operator, expr.args.map(subExpr => SUBRULE(exports.translateExpression, subExpr)));
        }
        if (F.isExpressionPatternOperation(expr)) {
            return AF.createExistenceExpression(expr.operator === 'notexists', SUBRULE(exports.translateGraphPattern, expr.args));
        }
        throw new Error(`Unknown expression: ${JSON.stringify(expr)}`);
    },
};
exports.translateGraphPattern = {
    name: 'translateGraphPattern',
    fun: ({ SUBRULE }) => ({ astFactory: F, algebraFactory: AF, useQuads }, pattern) => {
        // 18.2.2.1: Expand Syntax Forms -
        //    partly done by sparql parser, partly in this.translateTerm, and partly in BGP
        // https://www.w3.org/TR/sparql11-query/#sparqlExpandForms
        // https://www.w3.org/TR/sparql11-query/#QSynIRI
        if (F.isPatternBgp(pattern)) {
            return SUBRULE(exports.translateBgp, pattern);
        }
        // 18.2.2.6: Translate Graph Patterns - GroupOrUnionGraphPattern
        if (F.isPatternUnion(pattern)) {
            return AF.createUnion(pattern.patterns.map((group) => SUBRULE(exports.translateGraphPattern, group)));
        }
        // 18.2.2.6: Translate Graph Patterns - GraphGraphPattern
        if (F.isPatternGraph(pattern)) {
            // Sparql.js combines the group graph pattern and the graph itself in the same object.
            // We split here so the group graph pattern can be interpreted correctly.
            const group = F.patternGroup(pattern.patterns, pattern.loc);
            let result = SUBRULE(exports.translateGraphPattern, group);
            // Output depends on if we use quads or not
            if (useQuads) {
                result = SUBRULE(tripleAndQuad_js_1.recurseGraph, result, SUBRULE(general_js_1.translateTerm, pattern.name), undefined);
            }
            else {
                result = AF.createGraph(result, SUBRULE(general_js_1.translateTerm, pattern.name));
            }
            return result;
        }
        // 18.2.2.6: Translate Graph Patterns - InlineData
        if (F.isPatternValues(pattern)) {
            return SUBRULE(general_js_1.translateInlineData, pattern);
        }
        // 18.2.2.6: Translate Graph Patterns - SubSelect
        if (F.isQuerySelect(pattern)) {
            return SUBRULE(toAlgebra_js_1.translateQuery, pattern, useQuads, false);
        }
        // 18.2.2.6: Translate Graph Patterns - GroupGraphPattern
        if (F.isPatternGroup(pattern)) {
            // 18.2.2.2 - Collect FILTER Elements
            const filters = [];
            const nonfilters = [];
            for (const subPattern of pattern.patterns) {
                if (F.isPatternFilter(subPattern)) {
                    filters.push(subPattern);
                }
                else {
                    nonfilters.push(subPattern);
                }
            }
            // 18.2.2.6 - GroupGraphPattern
            let result = AF.createBgp([]);
            for (const pattern of nonfilters) {
                result = SUBRULE(exports.accumulateGroupGraphPattern, result, pattern);
            }
            // 18.2.2.7 - Filters of Group - translateExpression handles notExists negation.
            const expressions = filters.map(filter => SUBRULE(exports.translateExpression, filter.expression));
            if (expressions.length > 0) {
                let conjunction = expressions[0];
                for (const expression of expressions.slice(1)) {
                    conjunction = AF.createOperatorExpression('&&', [conjunction, expression]);
                }
                // One big filter applied on the group
                result = AF.createFilter(result, conjunction);
            }
            return result;
        }
        throw new Error(`Unexpected pattern: ${pattern.subType}`);
    },
};
/**
 * 18.2.2.1: Expand Syntax Forms: Flatten TripleCollection
 * 18.2.2.3: Translate Property Path Expressions
 * 18.2.2.4: Translate Property Path Patterns
 * 18.2.2.5: Translate Basic Graph Patterns
 */
exports.translateBgp = {
    name: 'translateBgp',
    fun: ({ SUBRULE }) => ({ astFactory: F, algebraFactory: AF }, bgp) => {
        let patterns = [];
        const joins = [];
        const flattenedTriples = [];
        SUBRULE(tripleAndQuad_js_1.translateBasicGraphPattern, bgp.triples, flattenedTriples);
        for (const triple of flattenedTriples) {
            if (F.isPathPure(triple.predicate)) {
                const smartType = triple;
                // TranslatePath returns a mix of Quads and Paths
                const path = SUBRULE(path_js_1.translatePath, smartType);
                for (const p of path) {
                    if (p.type === core_js_1.types.PATH) {
                        if (patterns.length > 0) {
                            joins.push(AF.createBgp(patterns));
                        }
                        patterns = [];
                        joins.push(p);
                    }
                    else {
                        patterns.push(p);
                    }
                }
            }
            else {
                patterns.push(SUBRULE(tripleAndQuad_js_1.translateQuad, triple));
            }
        }
        if (patterns.length > 0) {
            joins.push(AF.createBgp(patterns));
        }
        if (joins.length === 1) {
            return joins[0];
        }
        return AF.createJoin(joins);
    },
};
/**
 * 18.2.2.6 Translate Graph Patterns - GroupGraphPattern
 */
exports.accumulateGroupGraphPattern = {
    name: 'accumulateGroupGraphPattern',
    fun: ({ SUBRULE }) => ({ astFactory: F, algebraFactory: AF }, algebraOp, pattern) => {
        if (F.isPatternOptional(pattern)) {
            // Optional input needs to be interpreted as a group
            const groupAsAlgebra = SUBRULE(exports.translateGraphPattern, F.patternGroup(pattern.patterns, pattern.loc));
            if (groupAsAlgebra.type === core_js_1.types.FILTER) {
                return AF.createLeftJoin(algebraOp, groupAsAlgebra.input, groupAsAlgebra.expression);
            }
            return AF.createLeftJoin(algebraOp, groupAsAlgebra);
        }
        if (F.isPatternMinus(pattern)) {
            // Minus input needs to be interpreted as a group
            const groupAsAlgebra = SUBRULE(exports.translateGraphPattern, F.patternGroup(pattern.patterns, pattern.loc));
            return AF.createMinus(algebraOp, groupAsAlgebra);
        }
        if (F.isPatternBind(pattern)) {
            return AF.createExtend(algebraOp, SUBRULE(general_js_1.translateTerm, pattern.variable), SUBRULE(exports.translateExpression, pattern.expression));
        }
        if (F.isPatternService(pattern)) {
            // Transform to group so child-nodes get parsed correctly
            const group = F.patternGroup(pattern.patterns, pattern.loc);
            const A = AF.createService(SUBRULE(exports.translateGraphPattern, group), SUBRULE(general_js_1.translateTerm, pattern.name), pattern.silent);
            return SUBRULE(exports.simplifiedJoin, algebraOp, A);
        }
        const A = SUBRULE(exports.translateGraphPattern, pattern);
        return SUBRULE(exports.simplifiedJoin, algebraOp, A);
    },
};
exports.simplifiedJoin = {
    name: 'simplifiedJoin',
    fun: () => ({ algebraFactory: AF }, G, A) => {
        // Note: this is more simplification than requested in 18.2.2.8, but no reason not to do it.
        if (G.type === core_js_1.types.BGP && A.type === core_js_1.types.BGP) {
            G = AF.createBgp([...G.patterns, ...A.patterns]);
        }
        else if (G.type === core_js_1.types.BGP && G.patterns.length === 0) {
            // 18.2.2.8 (simplification)
            G = A;
        }
        else if (A.type === core_js_1.types.BGP && A.patterns.length === 0) {
            // Do nothing
        }
        else {
            G = AF.createJoin([G, A]);
        }
        return G;
    },
};
