// @traqula/algebra-transformations-1-1/dist/cjs/lib/toAlgebra/tripleAndQuad.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.translateQuad = exports.recurseGraph = exports.translateTripleNesting = exports.translateBasicGraphPattern = exports.translateTripleCollection = void 0;
const core_js_1 = require("./core.js");
const general_js_1 = require("./general.js");
exports.translateTripleCollection = {
    name: 'translateTripleCollection',
    fun: ({ SUBRULE }) => (_, collection, result) => {
        SUBRULE(exports.translateBasicGraphPattern, collection.triples, result);
    },
};
/**
 * When flattening, nested subject triples first, followed by nested object triples, lastly the current tripple.
 */
exports.translateBasicGraphPattern = {
    name: 'translateBasicGraphPattern',
    fun: ({ SUBRULE }) => ({ astFactory: F }, triples, result) => {
        for (const triple of triples) {
            if (F.isTripleCollection(triple)) {
                SUBRULE(exports.translateTripleCollection, triple, result);
            }
            else {
                SUBRULE(exports.translateTripleNesting, triple, result);
            }
        }
    },
};
/**
 * Generates: everything for the subject collection + everything for the object collection + itself
 */
exports.translateTripleNesting = {
    name: 'translateTripleNesting',
    fun: ({ SUBRULE }) => ({ astFactory: F }, triple, result) => {
        let subject;
        let predicate;
        let object;
        if (F.isTripleCollection(triple.subject)) {
            SUBRULE(exports.translateTripleCollection, triple.subject, result);
            subject = SUBRULE(general_js_1.translateTerm, triple.subject.identifier);
        }
        else {
            subject = SUBRULE(general_js_1.translateTerm, triple.subject);
        }
        if (F.isPathPure(triple.predicate)) {
            predicate = triple.predicate;
        }
        else {
            predicate = SUBRULE(general_js_1.translateTerm, triple.predicate);
        }
        if (F.isTripleCollection(triple.object)) {
            SUBRULE(exports.translateTripleCollection, triple.object, result);
            object = SUBRULE(general_js_1.translateTerm, triple.object.identifier);
        }
        else {
            object = SUBRULE(general_js_1.translateTerm, triple.object);
        }
        result.push({
            subject,
            predicate,
            object,
        });
    },
};
/**
 * Translate terms to be of some graph
 * @param c algebraContext
 * @param algOp algebra operation to translate
 * @param graph that should be assigned to the triples in algOp
 * @param replacement used for replacing shadowed variables.
 */
exports.recurseGraph = {
    name: 'recurseGraph',
    fun: ({ SUBRULE }) => (_, algOp, graph, replacement) => {
        if (algOp.type === core_js_1.types.GRAPH) {
            if (replacement) {
                // At this point we would lose track of the replacement which would result in incorrect results
                // This would indicate the library is not being used as intended though
                throw new Error('Recursing through nested GRAPH statements with a replacement is impossible.');
            }
            // In case there were nested GRAPH statements that were not recursed yet for some reason
            algOp = SUBRULE(exports.recurseGraph, algOp.input, algOp.name, undefined);
        }
        else if (algOp.type === core_js_1.types.SERVICE) {
            // Service blocks are not affected by enclosing GRAPH statements, so nothing is modified in this block.
            // See https://github.com/joachimvh/SPARQLAlgebra.js/pull/104#issuecomment-1838016303
        }
        else if (algOp.type === core_js_1.types.BGP) {
            algOp.patterns = algOp.patterns.map((quad) => {
                if (replacement) {
                    if (quad.subject.equals(graph)) {
                        quad.subject = replacement;
                    }
                    if (quad.predicate.equals(graph)) {
                        quad.predicate = replacement;
                    }
                    if (quad.object.equals(graph)) {
                        quad.object = replacement;
                    }
                }
                if (quad.graph.termType === 'DefaultGraph') {
                    quad.graph = graph;
                }
                return quad;
            });
        }
        else if (algOp.type === core_js_1.types.PATH) {
            if (replacement) {
                if (algOp.subject.equals(graph)) {
                    algOp.subject = replacement;
                }
                if (algOp.object.equals(graph)) {
                    algOp.object = replacement;
                }
            }
            if (algOp.graph.termType === 'DefaultGraph') {
                algOp.graph = graph;
            }
        }
        else if (algOp.type === core_js_1.types.PROJECT && !replacement) {
            // Need to replace variables in subqueries should the graph also be a variable of the same name
            // unless the subquery projects that variable
            if (!algOp.variables.some(v => v.equals(graph))) {
                replacement = SUBRULE(general_js_1.generateFreshVar);
            }
            algOp.input = SUBRULE(exports.recurseGraph, algOp.input, graph, replacement);
        }
        else if (algOp.type === core_js_1.types.EXTEND && !replacement) {
            // This can happen if the query extends an expression to the name of the graph
            // since the extend happens here there should be no further occurrences of this name
            // if there are it's the same situation as above
            if (algOp.variable.equals(graph)) {
                replacement = SUBRULE(general_js_1.generateFreshVar);
            }
            algOp.input = SUBRULE(exports.recurseGraph, algOp.input, graph, replacement);
        }
        else if (algOp.type === core_js_1.types.MINUS && graph.termType === 'Variable') {
            algOp.graphScopeVar = graph;
            algOp.input = [
                SUBRULE(exports.recurseGraph, algOp.input[0], graph, replacement),
                SUBRULE(exports.recurseGraph, algOp.input[1], graph, replacement),
            ];
        }
        else {
            for (const [key, value] of Object.entries(algOp)) {
                const castedKey = key;
                if (Array.isArray(value)) {
                    algOp[castedKey] = value.map((x) => SUBRULE(exports.recurseGraph, x, graph, replacement));
                }
                else if (core_js_1.typeVals.includes(value.type)) {
                    // Can't do instanceof on an interface
                    algOp[castedKey] = SUBRULE(exports.recurseGraph, value, graph, replacement);
                }
                else if (replacement && (0, core_js_1.isVariable)(value) && value.equals(graph)) {
                    algOp[castedKey] = replacement;
                }
            }
        }
        return algOp;
    },
};
exports.translateQuad = {
    name: 'translateQuad',
    fun: () => ({ astFactory: F, algebraFactory: AF }, quad) => {
        if (F.isPathPure(quad.predicate)) {
            throw new Error('Trying to translate property path to quad.');
        }
        // Graphs are needed here
        return AF.createPattern(quad.subject, quad.predicate, quad.object, quad.graph);
    },
};
