// @traqula/rules-sparql-1-1/dist/cjs/lib/grammar/dataSetClause.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.sourceSelector = exports.namedGraphClause = exports.usingClauseStar = exports.datasetClauseStar = exports.usingClause = exports.defaultGraphClause = exports.datasetClause = void 0;
exports.datasetClauseUsing = datasetClauseUsing;
exports.datasetClauseUsingStar = datasetClauseUsingStar;
const l = __importStar(require("../lexer/index.js"));
const literals_js_1 = require("./literals.js");
function datasetClauseUsing(name, token) {
    return {
        name,
        impl: ({ ACTION, SUBRULE, CONSUME, OR }) => (C) => {
            const start = CONSUME(token);
            return OR([
                { ALT: () => {
                        const iri = SUBRULE(exports.defaultGraphClause);
                        return ACTION(() => C.astFactory.wrap({ clauseType: 'default', value: iri }, C.astFactory.sourceLocation(start, iri)));
                    } },
                { ALT: () => {
                        const namedClause = SUBRULE(exports.namedGraphClause);
                        return ACTION(() => C.astFactory.wrap({
                            clauseType: 'named',
                            value: namedClause.val,
                        }, C.astFactory.sourceLocation(start, namedClause)));
                    } },
            ]);
        },
    };
}
/**
 * [[13]](https://www.w3.org/TR/sparql11-query/#rDatasetClause)
 */
exports.datasetClause = datasetClauseUsing('datasetClause', l.from);
/**
 * [[14]](https://www.w3.org/TR/sparql11-query/#rDefaultGraphClause)
 */
exports.defaultGraphClause = {
    name: 'defaultGraphClause',
    impl: ({ SUBRULE }) => () => SUBRULE(exports.sourceSelector),
};
/**
 * [[44]](https://www.w3.org/TR/sparql11-query/#rUsingClause)
 */
exports.usingClause = datasetClauseUsing('usingClause', l.usingClause);
function datasetClauseUsingStar(name, subRule, fromUsing) {
    return {
        name,
        impl: ({ ACTION, MANY, SUBRULE }) => (C) => {
            const clauses = [];
            MANY(() => {
                const clause = SUBRULE(subRule);
                clauses.push(clause);
            });
            return ACTION(() => C.astFactory.datasetClauses(clauses.map(clause => clause.val), C.astFactory.sourceLocation(...clauses)));
        },
        gImpl: ({ SUBRULE, PRINT_WORD }) => (ast, { astFactory: F }) => {
            for (const clause of ast.clauses) {
                F.printFilter(ast, () => PRINT_WORD(fromUsing));
                if (clause.clauseType === 'named') {
                    F.printFilter(ast, () => PRINT_WORD('NAMED'));
                }
                SUBRULE(literals_js_1.iri, clause.value);
            }
        },
    };
}
exports.datasetClauseStar = datasetClauseUsingStar('datasetClauses', exports.datasetClause, 'FROM');
exports.usingClauseStar = datasetClauseUsingStar('usingClauses', exports.usingClause, 'USING');
/**
 * [[15]](https://www.w3.org/TR/sparql11-query/#rNamedGraphClause)
 */
exports.namedGraphClause = {
    name: 'namedGraphClause',
    impl: ({ ACTION, SUBRULE, CONSUME }) => (C) => {
        const named = CONSUME(l.graph.named);
        const iri = SUBRULE(exports.sourceSelector);
        return ACTION(() => C.astFactory.wrap(iri, C.astFactory.sourceLocation(named, iri)));
    },
};
/**
 * [[16]](https://www.w3.org/TR/sparql11-query/#rSourceSelector)
 */
exports.sourceSelector = {
    name: 'sourceSelector',
    impl: ({ SUBRULE }) => () => SUBRULE(literals_js_1.iri),
};
