// @traqula/rules-sparql-1-1/dist/cjs/lib/grammar/general.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.graphTerm = exports.var_ = exports.varOrIri = exports.varOrTerm = exports.verb = exports.prefixDecl = exports.baseDecl = exports.prologue = void 0;
const l = __importStar(require("../lexer/index.js"));
const utils_js_1 = require("../utils.js");
const literals_js_1 = require("./literals.js");
/**
 * [[4]](https://www.w3.org/TR/sparql11-query/#rPrologue)
 */
exports.prologue = {
    name: 'prologue',
    impl: ({ SUBRULE, MANY, OR }) => () => {
        const result = [];
        MANY(() => OR([
            { ALT: () => result.push(SUBRULE(exports.baseDecl)) },
            // TODO: the [spec](https://www.w3.org/TR/sparql11-query/#iriRefs) says you cannot redefine prefixes.
            //  We might need to check this.
            { ALT: () => result.push(SUBRULE(exports.prefixDecl)) },
        ]));
        return result;
    },
    gImpl: ({ SUBRULE }) => (ast, { astFactory: F }) => {
        for (const context of ast) {
            if (F.isContextDefinitionBase(context)) {
                SUBRULE(exports.baseDecl, context);
            }
            else if (F.isContextDefinitionPrefix(context)) {
                SUBRULE(exports.prefixDecl, context);
            }
        }
    },
};
/**
 * Registers base IRI in the context and returns it.
 * [[5]](https://www.w3.org/TR/sparql11-query/#rBaseDecl)
 */
exports.baseDecl = {
    name: 'baseDecl',
    impl: ({ ACTION, CONSUME, SUBRULE }) => (C) => {
        const base = CONSUME(l.baseDecl);
        const val = SUBRULE(literals_js_1.iriFull);
        return ACTION(() => C.astFactory.contextDefinitionBase(C.astFactory.sourceLocation(base, val), val));
    },
    gImpl: ({ SUBRULE, PRINT_ON_EMPTY, NEW_LINE }) => (ast, { astFactory: F }) => {
        F.printFilter(ast, () => PRINT_ON_EMPTY('BASE '));
        SUBRULE(literals_js_1.iri, ast.value);
        F.printFilter(ast, () => NEW_LINE());
    },
};
/**
 * Registers prefix in the context and returns registered key-value-pair.
 * [[6]](https://www.w3.org/TR/sparql11-query/#rPrefixDecl)
 */
exports.prefixDecl = {
    name: 'prefixDecl',
    impl: ({ ACTION, CONSUME, SUBRULE }) => (C) => {
        const prefix = CONSUME(l.prefixDecl);
        const name = CONSUME(l.terminals.pNameNs).image.slice(0, -1);
        const value = SUBRULE(literals_js_1.iriFull);
        return ACTION(() => {
            C.prefixes[name] = value.value;
            return C.astFactory.contextDefinitionPrefix(C.astFactory.sourceLocation(prefix, value), name, value);
        });
    },
    gImpl: ({ SUBRULE, PRINT_ON_EMPTY, NEW_LINE }) => (ast, { astFactory: F }) => {
        F.printFilter(ast, () => {
            PRINT_ON_EMPTY('PREFIX ', `${ast.key}: `);
        });
        SUBRULE(literals_js_1.iri, ast.value);
        F.printFilter(ast, () => NEW_LINE());
    },
};
/**
 * [[78]](https://www.w3.org/TR/sparql11-query/#rVerb)
 */
exports.verb = {
    name: 'verb',
    impl: ({ SUBRULE, OR }) => () => OR([
        { ALT: () => SUBRULE(exports.varOrIri) },
        { ALT: () => SUBRULE(literals_js_1.verbA) },
    ]),
};
/**
 * [[106]](https://www.w3.org/TR/sparql11-query/#rVarOrTerm)
 */
exports.varOrTerm = {
    name: 'varOrTerm',
    impl: ({ SUBRULE, OR }) => C => OR([
        { GATE: () => C.parseMode.has('canParseVars'), ALT: () => SUBRULE(exports.var_) },
        { ALT: () => SUBRULE(exports.graphTerm) },
    ]),
    gImpl: ({ SUBRULE }) => (ast, { astFactory: F }) => {
        if (F.isTermVariable(ast)) {
            return SUBRULE(exports.var_, ast);
        }
        return SUBRULE(exports.graphTerm, ast);
    },
};
/**
 * [[107]](https://www.w3.org/TR/sparql11-query/#rVarOrIri)
 */
exports.varOrIri = {
    name: 'varOrIri',
    impl: ({ SUBRULE, OR }) => C => OR([
        { GATE: () => C.parseMode.has('canParseVars'), ALT: () => SUBRULE(exports.var_) },
        { ALT: () => SUBRULE(literals_js_1.iri) },
    ]),
};
/**
 * [[108]](https://www.w3.org/TR/sparql11-query/#rVar)
 */
exports.var_ = {
    name: 'var',
    impl: ({ ACTION, CONSUME, OR }) => (C) => {
        const varToken = OR([
            { ALT: () => CONSUME(l.terminals.var1) },
            { ALT: () => CONSUME(l.terminals.var2) },
        ]);
        return ACTION(() => C.astFactory.termVariable(varToken.image.slice(1), C.astFactory.sourceLocation(varToken)));
    },
    gImpl: ({ PRINT }) => (ast, { astFactory: F }) => {
        F.printFilter(ast, () => PRINT(`?${ast.value}`));
    },
};
/**
 * [[109]](https://www.w3.org/TR/sparql11-query/#rGraphTerm)
 */
exports.graphTerm = {
    name: 'graphTerm',
    impl: ({ ACTION, SUBRULE, CONSUME, OR }) => C => OR([
        { ALT: () => SUBRULE(literals_js_1.iri) },
        { ALT: () => SUBRULE(literals_js_1.rdfLiteral) },
        { ALT: () => SUBRULE(literals_js_1.numericLiteral) },
        { ALT: () => SUBRULE(literals_js_1.booleanLiteral) },
        { GATE: () => C.parseMode.has('canCreateBlankNodes'), ALT: () => SUBRULE(literals_js_1.blankNode) },
        { ALT: () => {
                const tokenNil = CONSUME(l.terminals.nil);
                return ACTION(() => C.astFactory.termNamed(C.astFactory.sourceLocation(tokenNil), utils_js_1.CommonIRIs.NIL));
            } },
    ]),
    gImpl: ({ SUBRULE }) => (ast, { astFactory: F }) => {
        if (F.isTermNamed(ast)) {
            SUBRULE(literals_js_1.iri, ast);
        }
        else if (F.isTermLiteral(ast)) {
            SUBRULE(literals_js_1.rdfLiteral, ast);
        }
        else if (F.isTermBlank(ast)) {
            SUBRULE(literals_js_1.blankNode, ast);
        }
    },
};
