// @traqula/rules-sparql-1-1/dist/cjs/lib/grammar/propertyPaths.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.pathOneInPropertySet = exports.pathNegatedPropertySet = exports.pathPrimary = exports.pathMod = exports.pathElt = exports.pathAlternative = exports.pathSequence = exports.pathEltOrInverse = exports.pathGenerator = exports.path = void 0;
exports.pathChainHelper = pathChainHelper;
const l = __importStar(require("../lexer/index.js"));
const literals_js_1 = require("./literals.js");
/**
 * [[88]](https://www.w3.org/TR/sparql11-query/#rPath)
 */
exports.path = {
    name: 'path',
    impl: ({ SUBRULE }) => () => SUBRULE(exports.pathAlternative),
};
exports.pathGenerator = {
    name: 'path',
    gImpl: ({ PRINT, SUBRULE }) => (ast, { astFactory: F }, braces = true) => {
        if (F.isTerm(ast) && F.isTermNamed(ast)) {
            SUBRULE(literals_js_1.iri, ast);
        }
        else {
            F.printFilter(ast, () => braces && PRINT('('));
            switch (ast.subType) {
                case '|':
                case '/': {
                    const [head, ...tail] = ast.items;
                    SUBRULE(exports.pathGenerator, head, braces);
                    for (const val of tail) {
                        F.printFilter(ast, () => PRINT(ast.subType));
                        SUBRULE(exports.pathGenerator, val, braces);
                    }
                    break;
                }
                case '^':
                    F.printFilter(ast, () => PRINT('^'));
                    SUBRULE(exports.pathGenerator, ast.items[0], braces);
                    break;
                case '?':
                case '*':
                case '+':
                    SUBRULE(exports.pathGenerator, ast.items[0], braces);
                    F.printFilter(ast, () => PRINT(ast.subType));
                    break;
                case '!':
                    F.printFilter(ast, () => PRINT('!'));
                    F.printFilter(ast, () => PRINT('('));
                    SUBRULE(exports.pathGenerator, ast.items[0], false);
                    F.printFilter(ast, () => PRINT(')'));
                    break;
            }
            F.printFilter(ast, () => braces && PRINT(')'));
        }
    },
};
function pathChainHelper(name, SEP, subType, subRule) {
    return {
        name,
        impl: ({ ACTION, CONSUME, SUBRULE1, SUBRULE2, MANY }) => (C) => {
            const head = SUBRULE1(subRule);
            let tailEnd = head;
            const tail = [];
            MANY(() => {
                CONSUME(SEP);
                tailEnd = SUBRULE2(subRule);
                tail.push(tailEnd);
            });
            return ACTION(() => tail.length === 0 ?
                head :
                C.astFactory.path(subType, [head, ...tail], C.astFactory.sourceLocation(head, tailEnd)));
        },
    };
}
/**
 * [[92]](https://www.w3.org/TR/sparql11-query/#rPathEltOrInverse)
 */
exports.pathEltOrInverse = {
    name: 'pathEltOrInverse',
    impl: ({ ACTION, CONSUME, SUBRULE1, SUBRULE2, OR }) => C => OR([
        { ALT: () => SUBRULE1(exports.pathElt) },
        { ALT: () => {
                const hat = CONSUME(l.symbols.hat);
                const item = SUBRULE2(exports.pathElt);
                return ACTION(() => C.astFactory.path('^', [item], C.astFactory.sourceLocation(hat, item)));
            } },
    ]),
};
/**
 * [[90]](https://www.w3.org/TR/sparql11-query/#rPathSequence)
 */
exports.pathSequence = pathChainHelper('pathSequence', l.symbols.slash, '/', exports.pathEltOrInverse);
/**
 * [[89]](https://www.w3.org/TR/sparql11-query/#rPathAlternative)
 */
exports.pathAlternative = pathChainHelper('pathAlternative', l.symbols.pipe, '|', exports.pathSequence);
/**
 * [[91]](https://www.w3.org/TR/sparql11-query/#rPathElt)
 */
exports.pathElt = {
    name: 'pathElt',
    impl: ({ ACTION, SUBRULE, OPTION }) => (C) => {
        const item = SUBRULE(exports.pathPrimary);
        const modification = OPTION(() => SUBRULE(exports.pathMod));
        return ACTION(() => modification === undefined ?
            item :
            C.astFactory.path(modification.image, [item], C.astFactory.sourceLocation(item, modification)));
    },
};
/**
 * [[93]](https://www.w3.org/TR/sparql11-query/#rPathMod)
 */
exports.pathMod = {
    name: 'pathMod',
    impl: ({ CONSUME, OR }) => () => OR([
        { ALT: () => CONSUME(l.symbols.question) },
        { ALT: () => CONSUME(l.symbols.star) },
        { ALT: () => CONSUME(l.symbols.opPlus) },
    ]),
};
/**
 * [[94]](https://www.w3.org/TR/sparql11-query/#rPathPrimary)
 */
exports.pathPrimary = {
    name: 'pathPrimary',
    impl: ({ SUBRULE, CONSUME, OR }) => () => OR([
        { ALT: () => SUBRULE(literals_js_1.iri) },
        { ALT: () => SUBRULE(literals_js_1.verbA) },
        { ALT: () => SUBRULE(exports.pathNegatedPropertySet) },
        { ALT: () => {
                CONSUME(l.symbols.LParen);
                const resRecursive = SUBRULE(exports.path);
                CONSUME(l.symbols.RParen);
                return resRecursive;
            } },
    ]),
};
/**
 * [[95]](https://www.w3.org/TR/sparql11-query/#rPathNegatedPropertySet)
 */
exports.pathNegatedPropertySet = {
    name: 'pathNegatedPropertySet',
    impl: ({ ACTION, CONSUME, SUBRULE1, SUBRULE2, SUBRULE3, OR, MANY }) => (C) => {
        const exclamation = CONSUME(l.symbols.exclamation);
        return OR([
            { ALT: () => {
                    const noAlternative = SUBRULE1(exports.pathOneInPropertySet);
                    return ACTION(() => C.astFactory.path('!', [noAlternative], C.astFactory.sourceLocation(exclamation, noAlternative)));
                } },
            { ALT: () => {
                    const open = CONSUME(l.symbols.LParen);
                    const head = SUBRULE2(exports.pathOneInPropertySet);
                    const tail = [];
                    MANY(() => {
                        CONSUME(l.symbols.pipe);
                        const item = SUBRULE3(exports.pathOneInPropertySet);
                        tail.push(item);
                    });
                    const close = CONSUME(l.symbols.RParen);
                    return ACTION(() => {
                        const F = C.astFactory;
                        if (tail.length === 0) {
                            return F.path('!', [head], F.sourceLocation(exclamation, close));
                        }
                        return F.path('!', [F.path('|', [head, ...tail], F.sourceLocation(open, close))], F.sourceLocation(exclamation, close));
                    });
                } },
        ]);
    },
};
/**
 * [[96]](https://www.w3.org/TR/sparql11-query/#rPathOneInPropertySet)
 */
exports.pathOneInPropertySet = {
    name: 'pathOneInPropertySet',
    impl: ({ ACTION, CONSUME, SUBRULE1, SUBRULE2, OR1, OR2 }) => C => OR1([
        { ALT: () => SUBRULE1(literals_js_1.iri) },
        { ALT: () => SUBRULE1(literals_js_1.verbA) },
        { ALT: () => {
                const hat = CONSUME(l.symbols.hat);
                const item = OR2([
                    { ALT: () => SUBRULE2(literals_js_1.iri) },
                    { ALT: () => SUBRULE2(literals_js_1.verbA) },
                ]);
                return ACTION(() => C.astFactory.path('^', [item], C.astFactory.sourceLocation(hat, item)));
            } },
    ]),
};
