// @traqula/algebra-transformations-1-2/dist/cjs/lib/toAlgebra12.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.inScopeVariables = exports.translateTripleNesting12 = exports.translateTripleCollection12 = exports.translateTerm12 = void 0;
const algebra_transformations_1_1_1 = require("@traqula/algebra-transformations-1-1");
const rules_sparql_1_2_1 = require("@traqula/rules-sparql-1-2");
const rdf_string_1 = require("rdf-string");
const reificationIri = 'http://www.w3.org/1999/02/22-rdf-syntax-ns#reifies';
exports.translateTerm12 = {
    name: 'translateTerm',
    fun: $ => (C, term) => {
        if (C.astFactory.isTermTriple(term)) {
            return C.dataFactory.quad($.SUBRULE(exports.translateTerm12, term.subject), $.SUBRULE(exports.translateTerm12, term.predicate), $.SUBRULE(exports.translateTerm12, term.object));
        }
        if (C.astFactory.isTermLiteral(term)) {
            if (!term.langOrIri) {
                return C.dataFactory.literal(term.value);
            }
            if (typeof term.langOrIri === 'object') {
                const iri = $.SUBRULE(algebra_transformations_1_1_1.translateNamed, term.langOrIri);
                return C.dataFactory.literal(term.value, iri);
            }
            const [lang, langDir] = term.langOrIri.split('--');
            if (langDir && (langDir === '' || langDir === 'ltr' || langDir === 'rtl')) {
                return C.dataFactory.literal(term.value, {
                    language: lang,
                    direction: langDir,
                });
            }
            return C.dataFactory.literal(term.value, term.langOrIri);
        }
        return algebra_transformations_1_1_1.translateTerm.fun($)(C, term);
    },
};
exports.translateTripleCollection12 = {
    name: 'translateTripleCollection',
    fun: s => (c, collection, result) => {
        if (c.astFactory.isTripleCollectionReifiedTriple(collection)) {
            const { SUBRULE } = s;
            const { dataFactory } = c;
            const identifier = SUBRULE(exports.translateTerm12, collection.identifier);
            SUBRULE(exports.translateTripleNesting12, collection.triples[0], result);
            // Given the rule for triple nesting, the triple itself is always generated last
            // (followed by annotations, but they are not allowed within reified triples
            const { subject, predicate, object } = result.pop();
            const asTerm = dataFactory.quad(subject, predicate, object);
            result.push({
                subject: identifier,
                predicate: dataFactory.namedNode(reificationIri),
                object: asTerm,
            });
        }
        else {
            algebra_transformations_1_1_1.translateTripleCollection.fun(s)(c, collection, result);
        }
    },
};
/**
 * Generates: everything for the subject collection + everything for the object collection + itself +
 *              the annotations triples
 */
exports.translateTripleNesting12 = {
    name: 'translateTripleNesting',
    fun: s => (c, triple, result) => {
        algebra_transformations_1_1_1.translateTripleNesting.fun(s)(c, triple, result);
        const SUBRULE = s.SUBRULE;
        const { subject, predicate, object } = result.at(-1);
        const { astFactory: F, dataFactory } = c;
        if (triple.annotations && triple.annotations.length > 0) {
            // Blocks know who identifies them. -> Can just add it to list of triples
            //  -> a blocks identifier only need to be registered when the identifier is not explicitly registered before.
            //  We register: <annotationNode, reifies, tripleNesting>
            // You cannot have an annotation on a triple that has paths
            const asTerm = dataFactory.quad(subject, predicate, object);
            const registered = new Set();
            for (const annotation of triple.annotations) {
                let subject;
                const tripleCollection = F.isTripleCollection(annotation);
                if (tripleCollection) {
                    subject = SUBRULE(exports.translateTerm12, annotation.identifier);
                }
                else {
                    subject = SUBRULE(exports.translateTerm12, annotation.val);
                }
                if (!registered.has((0, rdf_string_1.termToString)(subject))) {
                    result.push({
                        subject,
                        predicate: dataFactory.namedNode(reificationIri),
                        object: asTerm,
                    });
                }
                registered.add((0, rdf_string_1.termToString)(subject));
                if (tripleCollection) {
                    SUBRULE(exports.translateTripleCollection12, annotation, result);
                }
            }
        }
    },
};
/**
 * 18.2.1
 */
exports.inScopeVariables = {
    name: 'inScopeVariables',
    fun: () => (_, thingy) => {
        const vars = new Set();
        (0, rules_sparql_1_2_1.findPatternBoundedVars)(thingy, vars);
        return vars;
    },
};
